

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>besca.Import._read &mdash; BESCA 2.4.6+115.gb14a31c documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/plot_directive.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/sg_gallery.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/sg_gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/sg_gallery-dataframe.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/sg_gallery-rendered-html.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/clipboard.min.js"></script>
        <script src="../../../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../besca.html">besca</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html">tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>besca.Import._read</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for besca.Import._read</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>
<span class="kn">from</span> <span class="nn">scanpy</span> <span class="kn">import</span> <span class="n">read</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span><span class="p">,</span> <span class="n">issparse</span>

<span class="kn">from</span> <span class="nn">besca._helper</span> <span class="kn">import</span> <span class="n">convert_ensembl_to_symbol</span>

<span class="kn">import</span> <span class="nn">warnings</span>


<span class="k">def</span> <span class="nf">assert_filepath</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Asserts that the filepath contains the files required by read_mtx.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath: `str`</span>
<span class="sd">        filepath as string to the directory that is expected to contain at least</span>
<span class="sd">        following files: `matrix.mtx`, `genes.tsv`, and `barcodes.tsv` OR compressed</span>
<span class="sd">        versions of those files, i.e. `matrix.mtx.gz`, `genes.tsv.gz`, and `barcodes.tsv.gz`</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None if the filepath is valid, returns &#39;gz&#39; if valid and compressed files,</span>
<span class="sd">    otherwise an Exception will be raised</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">req_files</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;matrix.mtx&quot;</span><span class="p">,</span> <span class="s2">&quot;genes.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;barcodes.tsv&quot;</span><span class="p">]</span>
    <span class="n">req_files_exist</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">req_files</span><span class="p">]</span>
    <span class="n">valid</span> <span class="o">=</span> <span class="nb">all</span><span class="p">(</span><span class="n">req_files_exist</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">valid</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Check whether gzipped versions exist:</span>
    <span class="n">req_files</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;matrix.mtx.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;genes.tsv.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;barcodes.tsv.gz&quot;</span><span class="p">]</span>
    <span class="n">req_files_exist</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">req_files</span><span class="p">]</span>
    <span class="n">valid</span> <span class="o">=</span> <span class="nb">all</span><span class="p">(</span><span class="n">req_files_exist</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">valid</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;gz&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">exist</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">req_files_exist</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">exist</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
                    <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> is not found in &quot;</span>
                    <span class="s2">&quot;the given path `</span><span class="si">{}</span><span class="s2">`&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">req_files</span><span class="p">[</span><span class="n">ind</span><span class="p">],</span> <span class="n">filepath</span><span class="p">)</span>
                <span class="p">)</span>


<span class="k">def</span> <span class="nf">add_var_column</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">colname</span><span class="o">=</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="n">attempFix</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">colname</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">attempFix</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Creating empty </span><span class="si">{</span><span class="n">colname</span><span class="si">}</span><span class="s2"> column;  please check if </span><span class="si">{</span><span class="n">colname</span><span class="si">}</span><span class="s2"> is not in the index of adata.var&quot;</span>
            <span class="p">)</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">colname</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;N.A.&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;adata.var needs column named </span><span class="si">{colname}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adata</span>


<div class="viewcode-block" id="assert_adata"><a class="viewcode-back" href="../../../import/besca.Import.assert_adata.html#besca.Import.assert_adata">[docs]</a><span class="k">def</span> <span class="nf">assert_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="n">attempFix</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Asserts that an adata object is containing information needed for the besca pipeline to run and export information.</span>
<span class="sd">    This is particularly usefull when loading public data</span>
<span class="sd">    The parameter attempFix will try to fix the issue by itself.</span>
<span class="sd">    However, we advise the user to check by himself what is the leading problem.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: AnnData</span>
<span class="sd">    attempFix: `bool` if True will transform adata object to match requirements.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    returns an AnnData object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;CELL&quot;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">attempFix</span><span class="p">:</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;CELL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating columns CELL in adata.obs using adata.obs.index.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Required CELL columns in observations&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">==</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;CELL&quot;</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Required indexing of adata.obs by CELL column&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">attempFix</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Required count matrix to be sparse, X transformed to sparse&quot;</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;X transformation to sparse failed.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;adata.X needs to be sparse.&quot;</span><span class="p">)</span>
    <span class="c1"># checking adata.var concordance</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSEMBL&quot;</span><span class="p">]:</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">add_var_column</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">attempFix</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">el</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">for</span> <span class="n">el</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;In </span><span class="si">{x}</span><span class="s2"> non string values will create an issue for export&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adata</span></div>


<div class="viewcode-block" id="read_mtx"><a class="viewcode-back" href="../../../import/besca.Import.read_mtx.html#besca.Import.read_mtx">[docs]</a><span class="k">def</span> <span class="nf">read_mtx</span><span class="p">(</span>
    <span class="n">filepath</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_genes</span><span class="o">=</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">citeseq</span><span class="o">=</span><span class="kc">None</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read matrix.mtx, genes.tsv, barcodes.tsv to AnnData object.</span>
<span class="sd">    By specifiying an input folder this function reads the contained matrix.mtx,</span>
<span class="sd">    genes.tsv and barcodes.tsv files to an AnnData object. In case annotation = True</span>
<span class="sd">    it also adds the annotation contained in metadata.tsv to the object.</span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath: `str`</span>
<span class="sd">        filepath as string to the directory containg the matrix.mtx, genes.tsv,</span>
<span class="sd">        barcodes.tsv and if applicable metadata.tsv</span>
<span class="sd">    annotation: `bool` (default = True)</span>
<span class="sd">        boolian identifier if an annotation file is also located in the folder</span>
<span class="sd">        and should be added to the AnnData object</span>
<span class="sd">    use_genes: `str`</span>
<span class="sd">        either SYMBOL or ENSEMBL. Other genenames are not yet supported.</span>
<span class="sd">    species: `str` | default = &#39;human&#39;</span>
<span class="sd">        string specifying the species, only needs to be used when no Gene Symbols</span>
<span class="sd">        are supplied and you only have the ENSEMBLE gene ids to perform a lookup.</span>
<span class="sd">    citeseq: &#39;gex_only&#39; or &#39;citeseq_only&#39; or False or None | default = None</span>
<span class="sd">        string indicating if only gene expression values (gex_only) or only protein</span>
<span class="sd">        expression values (&#39;citeseq_only&#39;) or everything is read if None is specified</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    returns an AnnData object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gzfiles</span> <span class="o">=</span> <span class="n">assert_filepath</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gzfiles</span> <span class="o">==</span> <span class="s2">&quot;gz&quot;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading matrix.mtx.gz&quot;</span><span class="p">)</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;matrix.mtx.gz&quot;</span><span class="p">),</span> <span class="n">cache</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># transpose the data</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding cell barcodes&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;barcodes.tsv.gz&quot;</span><span class="p">),</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding genes&quot;</span><span class="p">)</span>
        <span class="n">var_anno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;genes.tsv.gz&quot;</span><span class="p">),</span>
            <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">t&quot;</span><span class="p">,</span>
            <span class="n">engine</span><span class="o">=</span><span class="s2">&quot;python&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading matrix.mtx&quot;</span><span class="p">)</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;matrix.mtx&quot;</span><span class="p">),</span> <span class="n">cache</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># transpose the data</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding cell barcodes&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;barcodes.tsv&quot;</span><span class="p">),</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding genes&quot;</span><span class="p">)</span>
        <span class="n">var_anno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;genes.tsv&quot;</span><span class="p">),</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">t&quot;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s2">&quot;python&quot;</span>
        <span class="p">)</span>

    <span class="n">symbols</span> <span class="o">=</span> <span class="n">var_anno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">ensembl_id</span> <span class="o">=</span> <span class="n">var_anno</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;ENSEMBL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ensembl_id</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">use_genes</span> <span class="o">==</span> <span class="s2">&quot;SYMBOL&quot;</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">symbols</span>

        <span class="c1"># make unique</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;making var_names unique&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">symbols</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">!=</span> <span class="n">ensembl_id</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding ENSEMBL gene ids to adata.var&quot;</span><span class="p">)</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">symbols</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">use_genes</span> <span class="o">==</span> <span class="s2">&quot;ENSEMBL&quot;</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">ensembl_id</span>

        <span class="k">if</span> <span class="n">symbols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ensembl_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="c1"># add symbols to object</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding symbols to adata.var&quot;</span><span class="p">)</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">symbols</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># lookup the corresponding Symbols</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">convert_ensembl_to_symbol</span><span class="p">(</span>
                <span class="n">ensembl_id</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">species</span><span class="o">=</span><span class="n">species</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s2">&quot;supplied unknown use_genes parameter&quot;</span><span class="p">)</span>

    <span class="n">check_response</span> <span class="o">=</span> <span class="n">check_data_for_citeseq</span><span class="p">(</span><span class="n">var_anno</span><span class="o">=</span><span class="n">var_anno</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">citeseq</span> <span class="o">==</span> <span class="s1">&#39;gex_only&#39;</span> <span class="ow">or</span> <span class="n">citeseq</span> <span class="o">==</span> <span class="s1">&#39;citeseq_only&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">check_response</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;genes.tsv&#39; does not include citeseq data or &#39;genes.tsv&#39; has invalid citeseq values.&quot;</span><span class="p">)</span>
        <span class="n">features</span> <span class="o">=</span> <span class="n">var_anno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="c1"># only extract the information we want</span>
        <span class="k">if</span> <span class="n">citeseq</span> <span class="o">==</span> <span class="s2">&quot;gex_only&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">citeseq</span> <span class="o">==</span> <span class="s2">&quot;citeseq_only&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">feature_type</span> <span class="o">==</span> <span class="s2">&quot;Antibody Capture&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">citeseq</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">check_response</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Citeseq data has been detected!, if you want to use it, please change citeseq parameter. Besca will continue without using the citeseq data&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">citeseq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">type</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span> <span class="ow">and</span> <span class="n">citeseq</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">check_response</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Citeseq data does not exists or is invalid!, change citeseq parameter to &#39;None&#39; if there is no citeseq data existing!&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;citeseq parameters has invalid value. Possible values: &#39;gex_only&#39;, &#39;citeseq_only&#39;, False or None.&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">annotation</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;adding annotation&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;metadata.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\\</span><span class="s2">t&quot;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s2">&quot;python&quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CELL&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CELL&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">adata</span></div>


<span class="k">def</span> <span class="nf">check_data_for_citeseq</span><span class="p">(</span><span class="n">var_anno</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read variable annotations and check if valid citeseq data is inlcluded</span>
<span class="sd">    ----------</span>
<span class="sd">    var_anno: `Pandas.Dataframe`</span>
<span class="sd">        A pandas dataframe, which includes the variable annotations.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    returns a boolean if citeseq data is existing</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">var_anno</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">:</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">var_anno</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;Gene Expression&#39;</span><span class="p">])</span> <span class="ow">or</span> \
            <span class="nb">set</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;Antibody Capture&#39;</span><span class="p">])</span> <span class="ow">or</span> \
            <span class="nb">set</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;Antibody Capture&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene Expression&#39;</span><span class="p">]):</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, Roche Pharma Research and Early Development,     Pharmaceutical Sciences, Roche Innovation Center Basel, Basel, Switzerland.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>