

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>besca.pp._fraction_pos &mdash; BESCA 2.3 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/gallery-dataframe.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../besca.html">besca</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials.html">tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>besca.pp._fraction_pos</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for besca.pp._fraction_pos</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">ndarray</span><span class="p">,</span> <span class="n">mean</span>

<div class="viewcode-block" id="frac_reads"><a class="viewcode-back" href="../../../preprocessing/besca.pp.frac_reads.html#besca.pp.frac_reads">[docs]</a><span class="k">def</span> <span class="nf">frac_reads</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Cacluate the fraction of reads being attributed to a specific gene.</span>
<span class="sd">    </span>
<span class="sd">    calculates the fraction of reads that are attributed to a specific gene over the entire dataset.</span>
<span class="sd">    This easily lets you identify genes which dominate all the reads in a dataset.</span>

<span class="sd">    parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: :class:`~anndata.AnnData`</span>
<span class="sd">        The annotated data matrix.</span>
<span class="sd">    threshold: `int` | default = 0</span>
<span class="sd">        Integer defining the value above which a gene is deemed as being expressed.</span>
<span class="sd">    </span>
<span class="sd">    returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        updates the:class:`~anndata.AnnData` object adding `frac_reads` to  `adata.var`.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    </span>
<span class="sd">    &gt;&gt;&gt; import besca as bc</span>
<span class="sd">    &gt;&gt;&gt; adata = bc.datasets.pbmc3k_raw()</span>
<span class="sd">    &gt;&gt;&gt; bc.pp.frac_reads(adata)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">ndarray</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>

        <span class="c1">#calculate the fraction of reads that are attributed to each gene</span>
        <span class="n">cum_reads_gene</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">adata_X</span><span class="p">)</span>
        <span class="n">frac_reads</span> <span class="o">=</span> <span class="n">cum_reads_gene</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">cum_reads_gene</span><span class="p">)</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;frac_reads&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">frac_reads</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">return</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;please pass an AnnData object as data&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="frac_pos"><a class="viewcode-back" href="../../../preprocessing/besca.pp.frac_pos.html#besca.pp.frac_pos">[docs]</a><span class="k">def</span> <span class="nf">frac_pos</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> 
             <span class="n">threshold</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate the fraction of cells positive for expression of a gene.</span>
<span class="sd">    </span>
<span class="sd">    Calculates for each gene the fraction of cells that show a gene expression</span>
<span class="sd">    above the indicated threshold. The fraction of positive cells is added to </span>
<span class="sd">    adata.var</span>

<span class="sd">    parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: :class:`~anndata.AnnData`</span>
<span class="sd">        The annotated data matrix.</span>
<span class="sd">    threshold: `int` | default = 0</span>
<span class="sd">        Integer defining the value above which a gene is deemed as being expressed.</span>
<span class="sd">    </span>
<span class="sd">    returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        updates the:class:`~anndata.AnnData` object adding `frac_pos` to  `adata.var`.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    </span>
<span class="sd">    &gt;&gt;&gt; import besca as bc</span>
<span class="sd">    &gt;&gt;&gt; adata = bc.datasets.pbmc3k_raw()</span>
<span class="sd">    &gt;&gt;&gt; bc.pp.frac_pos(adata)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">ndarray</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>

        <span class="c1">#calculate the percentage of cells that show an expression of the gene above the threshold</span>
        <span class="n">fraction_pos</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">adata_X</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">)</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1">#add calculated fraction positive to adata.var</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;frac_pos&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fraction_pos</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        
        <span class="k">return</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;please pass an AnnData object as data&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="mean_expr"><a class="viewcode-back" href="../../../preprocessing/besca.pp.mean_expr.html#besca.pp.mean_expr">[docs]</a><span class="k">def</span> <span class="nf">mean_expr</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate the mean expression of a gene.</span>
<span class="sd">    </span>
<span class="sd">    Calculates for each gene its mean expression in the adata object.</span>

<span class="sd">    parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: :class:`~anndata.AnnData`</span>
<span class="sd">        The annotated data matrix.</span>
<span class="sd">    </span>
<span class="sd">    returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        updates the:class:`~anndata.AnnData` object adding `mean` to  `adata.var`.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    </span>
<span class="sd">    &gt;&gt;&gt; import besca as bc</span>
<span class="sd">    &gt;&gt;&gt; adata = bc.datasets.pbmc3k_raw()</span>
<span class="sd">    &gt;&gt;&gt; bc.pp.mean_expr(adata)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">ndarray</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">adata_X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>

        <span class="c1">#calculate the percentage of cells that show an expression of the gene above the threshold</span>
        <span class="n">means</span> <span class="o">=</span> <span class="n">mean</span><span class="p">(</span><span class="n">adata_X</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

        <span class="c1">#add calculated fraction positive to adata.var</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        
        <span class="k">return</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;please pass an AnnData object as data&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="top_counts_genes"><a class="viewcode-back" href="../../../preprocessing/besca.pp.top_counts_genes.html#besca.pp.top_counts_genes">[docs]</a><span class="k">def</span> <span class="nf">top_counts_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Give out the genes that contribute the largest fraction to the total UMI counts.</span>

<span class="sd">    Calculates the genes which contribute the largest fraction of UMI counts to the total</span>
<span class="sd">     UMI counts (i.e. the most abundant transcripts).</span>

<span class="sd">    parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: :class:`~anndata.AnnData`</span>
<span class="sd">        The annotated data matrix.</span>
<span class="sd">    top_n: `int` | default = 10</span>
<span class="sd">        Integer defining the number of entries to return. If None then the entire table is returned.</span>

<span class="sd">    returns</span>
<span class="sd">    -------</span>
<span class="sd">    DataFrame</span>
<span class="sd">        pandas DataFrame containing the top_n genes </span>
<span class="sd">    </span>
<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; import besca as bc</span>
<span class="sd">    &gt;&gt;&gt; adata = bc.datasets.pbmc3k_raw()</span>
<span class="sd">    &gt;&gt;&gt; bc.pp.top_counts_genes(adata)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>   
        <span class="k">if</span> <span class="s1">&#39;frac_reads&#39;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>

            <span class="c1">#generate a table sorted by frac_pos (descending order)</span>
            <span class="nb">sorted</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;frac_reads&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">top_n</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;calculating frac_reads&#39;</span><span class="p">)</span>

            <span class="c1">#calculate fraciton positive for a gene</span>
            <span class="n">bdata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">frac_reads</span><span class="p">(</span><span class="n">bdata</span><span class="p">)</span>

            <span class="c1">#generate a table sorted by frac_pos (descending order)</span>
            <span class="nb">sorted</span> <span class="o">=</span> <span class="n">bdata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;frac_reads&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">top_n</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;please pass an AnnData object as data&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="top_expressed_genes"><a class="viewcode-back" href="../../../preprocessing/besca.pp.top_expressed_genes.html#besca.pp.top_expressed_genes">[docs]</a><span class="k">def</span> <span class="nf">top_expressed_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Give out the genes most frequently expressed in cells.</span>

<span class="sd">    Calculates the genes most frequently expressed in cells and returns a table</span>
<span class="sd">    containing the top_n entries.</span>

<span class="sd">    parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata: :class:`~anndata.AnnData`</span>
<span class="sd">        The annotated data matrix.</span>
<span class="sd">    top_n: `int` | default = 10</span>
<span class="sd">        Integer defining the number of entries to return. If None then the entire table is returned.</span>

<span class="sd">    returns</span>
<span class="sd">    -------</span>
<span class="sd">    DataFrame</span>
<span class="sd">        pandas DataFrame containing the top_n genes </span>
<span class="sd">    </span>
<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; import besca as bc</span>
<span class="sd">    &gt;&gt;&gt; adata = bc.datasets.pbmc3k_raw()</span>
<span class="sd">    &gt;&gt;&gt; bc.pp.top_expressed_genes(adata)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>   
        <span class="k">if</span> <span class="s1">&#39;frac_pos&#39;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>

            <span class="c1">#generate a table sorted by frac_pos (descending order)</span>
            <span class="nb">sorted</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;frac_pos&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">top_n</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;calculating frac_pos&#39;</span><span class="p">)</span>

            <span class="c1">#calculate fraciton positive for a gene</span>
            <span class="n">bdata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">frac_pos</span><span class="p">(</span><span class="n">bdata</span><span class="p">)</span>

            <span class="c1">#generate a table sorted by frac_pos (descending order)</span>
            <span class="nb">sorted</span> <span class="o">=</span> <span class="n">bdata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;frac_pos&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">top_n</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="nb">sorted</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;please pass an AnnData object as data&#39;</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, BEDA

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>