

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Automated Annotation Workflow &mdash; BESCA 2.4.6+115.gb14a31c documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../_static/plot_directive.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-dataframe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-rendered-html.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="besca standard pipeline" href="../besca_standard_pipeline.html" />
    <link rel="prev" title="notebook 3 - batch correction" href="notebook3_batch_correction.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../besca.html">besca</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="notebook1_data_processing_pbmc3k.html">notebook 1 - introduction and data processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebook2_celltype_annotation_pbmc3k.html">notebook 2 - celltype annotation and beyond</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebook3_batch_correction.html">notebook 3 - batch correction</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Automated Annotation Workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Parameter-specification">Parameter specification</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Data-loading">Data loading</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Read-in-all-training-sets-and-the-testing-set.">Read in all training sets and the testing set.</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Correct-datasets-(e.g. using-scanorama)-and-merge-training-datasets">Correct datasets (e.g. using scanorama) and merge training datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Train-the-classifier">Train the classifier</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Prediction">Prediction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Output">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Let’s-try-another-classifier-(SVM)">Let’s try another classifier (SVM)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Let’s-try-with-a-specified-gene-set">Let’s try with a specified gene set</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Done!">Done!</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#single-cell-sequencing-general-tutorials">single cell sequencing general tutorials</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#single-cell-auto-annot-tutorial-for-cell-type-annotation">single cell auto_annot tutorial for cell type annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#bescape-cell-deconvolution-tutorial">Bescape: cell deconvolution tutorial</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">tutorials</a> &raquo;</li>
        
      <li>Automated Annotation Workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tutorials/auto_annot_tutorial.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Automated-Annotation-Workflow">
<h1>Automated Annotation Workflow<a class="headerlink" href="#Automated-Annotation-Workflow" title="Permalink to this heading">¶</a></h1>
<p>This workflow uses the auto_annot tools from besca to newly annotate a scRNAseq dataset based on one or more preannotated datasets. Ideally, these datasets come from a similar tissue and condition. If multiple training datasets are used, the performance is dependent on all dataset being annotated to the same resolution with broadly similar cell types.</p>
<p>We use supervised machine learning methods to annotate each individual cell utilizing methods like support vector machines (SVM) or logistic regression.</p>
<p>First, the traning dataset(s) and the testing dataset are loaded from h5ad files or made available as adata objects. Next, the training and testing datasets are corrected using scanorama, and the training datasets are then merged into one anndata object. Then, the classifier is trained utilizing the merged training data. Finally, the classifier is applied to the testing dataset to predict the cell types. If the testing dataset is already annotated (to test the algorithm), a report including
confusion matrices can be generated.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">besca</span> <span class="k">as</span> <span class="nn">bc</span>
<br/></pre></div>
</div>
</div>
<section id="Parameter-specification">
<h2>Parameter specification<a class="headerlink" href="#Parameter-specification" title="Permalink to this heading">¶</a></h2>
<p>Give your analysis a name.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;auto_annot_tutorial&quot;</span>  <span class="c1"># The analysis name will be used to name the output files</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>Specify column name of celltype annotation you want to train on.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype</span> <span class="o">=</span> <span class="s2">&quot;dblabel&quot;</span>  <span class="c1"># This needs to be a column in the .obs of the training datasets (and test dataset if you want to generate a report)</span>
<br/></pre></div>
</div>
</div>
<p>Choose a method:</p>
<ul class="simple">
<li><p>linear: Support Vector Machine with Linear Kernel</p></li>
<li><p>sgd: Support Vector Machine with Linear Kernel using Stochastic Gradient Descent</p></li>
<li><p>rbf: Support Vector Machine with radial basis function kernel. Very time intensive, use only on small datasets.</p></li>
<li><p>logistic_regression: Standard logistic classifier iwth multinomial loss.</p></li>
<li><p>logistic_regression_ovr: Logistic Regression with one versus rest classification.</p></li>
<li><p>logistic_regression_elastic: Logistic Regression with elastic loss, cross validates among multiple l1 ratios.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;logistic_regression&quot;</span>
<br/></pre></div>
</div>
</div>
<p>Specify merge method. Needs to be either scanorama or naive.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">merge</span> <span class="o">=</span> <span class="s2">&quot;scanorama&quot;</span>  <span class="c1"># We recommend to use scanorama here</span>
<br/></pre></div>
</div>
</div>
<p>Decide if you want to use the raw format or highly variable genes. Raw increases computational time and does not necessarily improve predictions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># We recommend to use False here</span>
<br/></pre></div>
</div>
</div>
<p>You can choose to only consider a subset of genes from a signature set or use all genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes_to_use</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>  <span class="c1"># We suggest to use all here, but the runtime is strongly improved if you select an appropriate gene set</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="Data-loading">
<h2>Data loading<a class="headerlink" href="#Data-loading" title="Permalink to this heading">¶</a></h2>
<section id="Read-in-all-training-sets-and-the-testing-set.">
<h3>Read in all training sets and the testing set.<a class="headerlink" href="#Read-in-all-training-sets-and-the-testing-set." title="Permalink to this heading">¶</a></h3>
<p>We will use a publicly available PBMC dataset, including ~3000 cells, as testing dataset. For the training dataset we use another PBMC dataset that is delivered with besca. The training datasets used are from:</p>
<p>Granja JM, Klemm S, McGinnis LM, et al. Single-cell multiomic analysis identifies regulatory programs in mixed-phenotype acute leukemia. Nat Biotechnol. 2019;37(12):1458-1465. doi:10.1038/s41587-019-0332-7</p>
<p>and</p>
<p>Kotliarov Y, Sparks R, Martins AJ, et al. Broad immune activation underlies shared set point signatures for vaccine responsiveness in healthy individuals and disease activity in patients with lupus. Nat Med. 2020;26(4):618-629. doi:10.1038/s41591-020-0769-8</p>
<p>A detailed analysis of the annotation performed below can be found in figure 3 of the publication on besca.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k_processed</span><span class="p">()</span>
<span class="p">)</span>  <span class="c1"># Dataset to be annotated (it&#39;s already annotated in this case for testing)</span>
<span class="n">adata_test_orig</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata_test</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="p">)</span>  <span class="c1"># Make a copy of the testing data, which will be annotated in .obs, but not further modified</span>
<span class="n">adata_train_list</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Granja2019_processed</span><span class="p">(),</span>
    <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Kotliarov2020_processed</span><span class="p">(),</span>
<span class="p">]</span>  <span class="c1"># List of annotated training datasets</span>
<br/></pre></div>
</div>
</div>
<p>You can also load your own dataset from the file system. Specify folders where .h5ad files are found and their names. The datasets that are already annotated and should be used for training. If you only use one dataset please use list of one. These load functions are useful in particular when the datasets have been processed with the besca standard pipeline. Alternatively, datasets can be loaded in any fashion, if more suitable.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># test_dataset_path = &#39;/path/to/test/dataset/folder&#39;</span>
<span class="c1"># test_dataset = &#39;testdataset.h5ad&#39;</span>
<span class="c1">#</span>
<span class="c1"># train_dataset_paths = [&#39;/path/to/train/dataset/folder1&#39;, &#39;/path/to/train/dataset/folder2&#39;]</span>
<span class="c1"># train_datasets = [&#39;traindataset1.h5ad&#39;, &#39;traindataset2.h5ad&#39;]</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># adata_train_list, adata_test, adata_test_orig = bc.tl.auto_annot.read_data(train_paths = train_dataset_paths, train_datasets= train_datasets, test_path=  test_dataset_path, test_dataset= test_dataset, use_raw = use_raw)</span>
<br/></pre></div>
</div>
</div>
<p>One training dataset’s label column is not named “dblabel”. We need to create a column so that all datasets used contain a column that are named according to the previously specified variable celltype and that contain the labels.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dblabel&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype3&quot;</span><span class="p">]</span>
<br/></pre></div>
</div>
</div>
<p>In this case the testing dataset is already annotated to demonstrate the methodology. All datasets adhere to the same naming convention.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_test</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;naive thymus-derived CD8-positive, alpha-beta..., &#39;naive B cell&#39;, &#39;central memory CD4-positive, alpha-beta T cell&#39;, &#39;classical monocyte&#39;, &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T..., &#39;non-classical monocyte&#39;, &#39;naive thymus-derived CD4-positive, alpha-beta..., &#39;CD8-positive, alpha-beta cytotoxic T cell&#39;, &#39;cytotoxic CD56-dim natural killer cell&#39;, &#39;CD1c-positive myeloid dendritic cell&#39;]
Categories (10, object): [&#39;CD1c-positive myeloid dendritic cell&#39;, &#39;CD8-positive, alpha-beta cytotoxic T cell&#39;, &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T..., &#39;central memory CD4-positive, alpha-beta T cell&#39;, ..., &#39;naive B cell&#39;, &#39;naive thymus-derived CD4-positive, alpha-beta..., &#39;naive thymus-derived CD8-positive, alpha-beta..., &#39;non-classical monocyte&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;naive thymus-derived CD4-positive, alpha-beta..., &#39;classical monocyte&#39;, &#39;naive B cell&#39;, &#39;lymphocyte of B lineage&#39;, &#39;naive thymus-derived CD8-positive, alpha-beta..., ..., &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T..., &#39;hematopoietic multipotent progenitor cell&#39;, &#39;myeloid leukocyte&#39;, &#39;basophil&#39;, &#39;plasma cell&#39;]
Length: 25
Categories (25, object): [&#39;naive thymus-derived CD4-positive, alpha-beta..., &#39;naive thymus-derived CD8-positive, alpha-beta..., &#39;naive B cell&#39;, &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T..., ..., &#39;common lymphoid progenitor&#39;, &#39;basophil&#39;, &#39;plasma cell&#39;, &#39;erythrocyte&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;cytotoxic CD56-dim natural killer cell&#39;, &#39;naive thymus-derived CD8-positive, alpha-beta..., &#39;naive thymus-derived CD4-positive, alpha-beta..., &#39;classical monocyte&#39;, &#39;CD8-positive, alpha-beta cytotoxic T cell&#39;, ..., &#39;regulatory T cell&#39;, &#39;CD1c-positive myeloid dendritic cell&#39;, &#39;plasmacytoid dendritic cell&#39;, &#39;erythrocyte&#39;, &#39;plasma cell&#39;]
Length: 14
Categories (14, object): [&#39;cytotoxic CD56-dim natural killer cell&#39;, &#39;naive B cell&#39;, &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T..., &#39;memory B cell&#39;, ..., &#39;plasmacytoid dendritic cell&#39;, &#39;erythrocyte&#39;, &#39;plasma cell&#39;, &#39;naive thymus-derived CD8-positive, alpha-beta...]
</pre></div></div>
</div>
</section>
</section>
<section id="Correct-datasets-(e.g. using-scanorama)-and-merge-training-datasets">
<h2>Correct datasets (e.g. using scanorama) and merge training datasets<a class="headerlink" href="#Correct-datasets-(e.g. using-scanorama)-and-merge-training-datasets" title="Permalink to this heading">¶</a></h2>
<p>This function merges training datasets, removes unwanted genes, and if scanorama is used corrects for datasets.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span>
    <span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="o">=</span><span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="n">merge</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
merging with scanorama
using scanorama rn
Found 207 genes among all datasets
[[0.         0.69861833 0.47963259]
 [0.         0.         0.98881789]
 [0.         0.         0.        ]]
Processing datasets (1, 2)
Processing datasets (0, 1)
Processing datasets (0, 2)
integrating training set
</pre></div></div>
</div>
</section>
<section id="Train-the-classifier">
<h2>Train the classifier<a class="headerlink" href="#Train-the-classifier" title="Permalink to this heading">¶</a></h2>
<p>The returned scaler is fitted on the training dataset (to zero mean and scaled to unit variance). The scaling will then be applied to the counts in the testing dataset and then the classifier is applied to the scaled testing dataset (see next step, adata_predict()). This function will run multiple jobs in parallel if if logistic regression was specified as method.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.12039D+05    |proj g|=  2.86381D+04
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.11969D+05    |proj g|=  2.70303D+04
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.11946D+05    |proj g|=  2.74976D+04
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.12051D+05    |proj g|=  2.70821D+04
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  2.11957D+05    |proj g|=  2.71735D+04

At iterate   50    f=  1.04971D+05    |proj g|=  2.73525D+01

At iterate   50    f=  1.03637D+05    |proj g|=  8.64883D+00

At iterate   50    f=  1.03011D+05    |proj g|=  8.89155D+00

At iterate   50    f=  1.03004D+05    |proj g|=  5.90025D+00

At iterate   50    f=  1.04246D+05    |proj g|=  7.64672D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200     70     73      1     0     0   1.162D+00   1.030D+05
  F =   103010.68989000401

CONVERGENCE: REL_REDUCTION_OF_F_&lt;=_FACTR*EPSMCH
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  7.74054D+04    |proj g|=  2.54100D+03

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200     72     76      1     0     0   1.197D+00   1.050D+05
  F =   104971.11070892689

CONVERGENCE: REL_REDUCTION_OF_F_&lt;=_FACTR*EPSMCH
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  7.96649D+04    |proj g|=  2.47790D+03

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200     74     78      1     0     0   6.220D-01   1.036D+05
  F =   103636.52996705240

CONVERGENCE: REL_REDUCTION_OF_F_&lt;=_FACTR*EPSMCH
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  7.83556D+04    |proj g|=  2.44585D+03

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200     77     82      1     0     0   6.687D-01   1.030D+05
  F =   103003.76722693436

CONVERGENCE: REL_REDUCTION_OF_F_&lt;=_FACTR*EPSMCH
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  7.76244D+04    |proj g|=  2.55120D+03

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200     80     85      1     0     0   8.110D-01   1.042D+05
  F =   104245.97290945785

CONVERGENCE: REL_REDUCTION_OF_F_&lt;=_FACTR*EPSMCH
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  7.90204D+04    |proj g|=  2.50165D+03

At iterate   50    f=  5.93638D+04    |proj g|=  2.00635D+01

At iterate   50    f=  6.15150D+04    |proj g|=  3.03599D+01

At iterate   50    f=  6.04466D+04    |proj g|=  1.57444D+01

At iterate   50    f=  5.98289D+04    |proj g|=  2.91009D+01

At iterate   50    f=  6.11011D+04    |proj g|=  2.25134D+01

At iterate  100    f=  5.93575D+04    |proj g|=  4.64568D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    104      1     0     0   4.646D+00   5.936D+04
  F =   59357.500019558975

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  4.73539D+04    |proj g|=  5.57261D+02

At iterate  100    f=  6.15099D+04    |proj g|=  1.67961D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.680D+00   6.151D+04
  F =   61509.900886338968

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  4.94544D+04    |proj g|=  6.00582D+02

At iterate  100    f=  6.04395D+04    |proj g|=  1.89094D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.891D+00   6.044D+04
  F =   60439.492975334550

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  4.85205D+04    |proj g|=  5.86978D+02

At iterate  100    f=  5.98230D+04    |proj g|=  1.48425D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.484D+00   5.982D+04
  F =   59822.961628412340

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  4.79571D+04    |proj g|=  5.79302D+02

At iterate  100    f=  6.10936D+04    |proj g|=  2.97156D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   2.972D+00   6.109D+04
  F =   61093.551063188512

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  4.91858D+04    |proj g|=  5.88940D+02

At iterate   50    f=  4.11176D+04    |proj g|=  2.70626D+01

At iterate   50    f=  4.31390D+04    |proj g|=  3.15829D+01

At iterate   50    f=  4.23066D+04    |proj g|=  3.96669D+01

At iterate   50    f=  4.17778D+04    |proj g|=  2.30889D+01

At iterate   50    f=  4.30381D+04    |proj g|=  2.38411D+01

At iterate  100    f=  4.10888D+04    |proj g|=  7.57278D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    104      1     0     0   7.573D+00   4.109D+04
  F =   41088.792765115402

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.64715D+04    |proj g|=  1.14671D+02

At iterate  100    f=  4.22764D+04    |proj g|=  1.94224D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.942D+01   4.228D+04
  F =   42276.428932985196

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  4.31078D+04    |proj g|=  2.42772D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   2.428D+01   4.311D+04
  F =   43107.796510953172

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.76858D+04    |proj g|=  1.23967D+02
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.84692D+04    |proj g|=  1.18329D+02

At iterate  100    f=  4.17517D+04    |proj g|=  4.54061D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    105      1     0     0   4.541D+01   4.175D+04
  F =   41751.728515726456

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.71857D+04    |proj g|=  1.24753D+02

At iterate  100    f=  4.30019D+04    |proj g|=  1.31606D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    103      1     0     0   1.316D+01   4.300D+04
  F =   43001.948646084813

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.84556D+04    |proj g|=  1.19735D+02

At iterate   50    f=  3.43179D+04    |proj g|=  1.61990D+01

At iterate   50    f=  3.55724D+04    |proj g|=  5.30592D+01

At iterate   50    f=  3.63519D+04    |proj g|=  2.47038D+01

At iterate   50    f=  3.50436D+04    |proj g|=  3.81051D+01

At iterate   50    f=  3.63920D+04    |proj g|=  5.06944D+01

At iterate  100    f=  3.42672D+04    |proj g|=  1.54011D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    105      1     0     0   1.540D+01   3.427D+04
  F =   34267.225020703758

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.26307D+04    |proj g|=  2.29930D+01

At iterate  100    f=  3.62979D+04    |proj g|=  6.97876D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   6.979D+01   3.630D+04
  F =   36297.920086924394

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.55243D+04    |proj g|=  2.52911D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   2.529D+01   3.552D+04
  F =   35524.276263567997

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.46728D+04    |proj g|=  7.72150D+01

At iterate  100    f=  3.49969D+04    |proj g|=  9.13133D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    104      1     0     0   9.131D+00   3.500D+04
  F =   34996.943723823075

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.39185D+04    |proj g|=  3.31235D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.33719D+04    |proj g|=  2.34712D+01

At iterate  100    f=  3.63334D+04    |proj g|=  1.43529D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.435D+01   3.633D+04
  F =   36333.443243589172

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.47128D+04    |proj g|=  2.46425D+01

At iterate   50    f=  3.20623D+04    |proj g|=  5.45743D+01

At iterate   50    f=  3.41204D+04    |proj g|=  3.54927D+01

At iterate   50    f=  3.33592D+04    |proj g|=  3.47271D+01

At iterate   50    f=  3.27833D+04    |proj g|=  5.03492D+01

At iterate   50    f=  3.41880D+04    |proj g|=  8.07172D+01

At iterate  100    f=  3.19593D+04    |proj g|=  5.11212D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    110      1     0     0   5.112D+01   3.196D+04
  F =   31959.310456641466

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.14746D+04    |proj g|=  5.22624D+01

At iterate  100    f=  3.40062D+04    |proj g|=  1.23469D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.235D+01   3.401D+04
  F =   34006.179329086517

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.35020D+04    |proj g|=  1.28782D+01

At iterate  100    f=  3.32580D+04    |proj g|=  2.06409D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   2.064D+01   3.326D+04
  F =   33258.003907179125

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.26660D+04    |proj g|=  1.49031D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   1.490D+01   3.267D+04
  F =   32666.015817912816

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.27945D+04    |proj g|=  2.22048D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.21525D+04    |proj g|=  1.54769D+01

At iterate  100    f=  3.40856D+04    |proj g|=  2.61951D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   2.620D+01   3.409D+04
  F =   34085.608687699583

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.36314D+04    |proj g|=  2.70166D+01

At iterate   50    f=  3.13759D+04    |proj g|=  5.13549D+01

At iterate   50    f=  3.26863D+04    |proj g|=  4.69096D+01

At iterate   50    f=  3.34000D+04    |proj g|=  2.70486D+01

At iterate   50    f=  3.20391D+04    |proj g|=  1.90035D+01

At iterate   50    f=  3.35226D+04    |proj g|=  9.19913D+01

At iterate  100    f=  3.12623D+04    |proj g|=  2.61464D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    109      1     0     0   2.615D+01   3.126D+04
  F =   31262.281830163720

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.11541D+04    |proj g|=  2.60008D+01

At iterate  100    f=  3.25599D+04    |proj g|=  4.23087D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   4.231D+01   3.256D+04
  F =   32559.924042316710

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.24459D+04    |proj g|=  4.21871D+01

At iterate  100    f=  3.32669D+04    |proj g|=  1.81318D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.813D+01   3.327D+04
  F =   33266.879532518455

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.31391D+04    |proj g|=  1.78562D+01

At iterate  100    f=  3.18701D+04    |proj g|=  2.01354D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   2.014D+01   3.187D+04
  F =   31870.144245662003

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.17202D+04    |proj g|=  2.02726D+01

At iterate  100    f=  3.33875D+04    |proj g|=  3.87361D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    105      1     0     0   3.874D+01   3.339D+04
  F =   33387.501525823609

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.32710D+04    |proj g|=  3.89076D+01

At iterate   50    f=  3.24120D+04    |proj g|=  1.68045D+01

At iterate   50    f=  3.11244D+04    |proj g|=  2.11394D+01

At iterate   50    f=  3.31137D+04    |proj g|=  1.36011D+01

At iterate   50    f=  3.16790D+04    |proj g|=  1.49819D+01

At iterate   50    f=  3.32350D+04    |proj g|=  2.77797D+01

At iterate  100    f=  3.10458D+04    |proj g|=  5.27721D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   5.277D+01   3.105D+04
  F =   31045.808935526016

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.23504D+04    |proj g|=  1.68157D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.682D+01   3.235D+04
  F =   32350.360595808739

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.10268D+04    |proj g|=  5.27721D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.23327D+04    |proj g|=  1.68403D+01

At iterate  100    f=  3.16356D+04    |proj g|=  2.32266D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   2.323D+01   3.164D+04
  F =   31635.616571226965

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.30666D+04    |proj g|=  2.88477D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   2.885D+01   3.307D+04
  F =   33066.577323674304

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.16146D+04    |proj g|=  2.32253D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.30481D+04    |proj g|=  2.88754D+01

At iterate  100    f=  3.31910D+04    |proj g|=  1.95868D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.959D+01   3.319D+04
  F =   33190.978410728567

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.31747D+04    |proj g|=  1.95730D+01

At iterate   50    f=  3.23028D+04    |proj g|=  1.67149D+01

At iterate   50    f=  3.10050D+04    |proj g|=  6.82002D+00

At iterate   50    f=  3.15900D+04    |proj g|=  1.27052D+01

At iterate   50    f=  3.30243D+04    |proj g|=  8.12354D+00

At iterate   50    f=  3.31546D+04    |proj g|=  2.98043D+01

At iterate  100    f=  3.22851D+04    |proj g|=  2.58396D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   2.584D+01   3.229D+04
  F =   32285.126911492029

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.22824D+04    |proj g|=  2.58431D+01

At iterate  100    f=  3.09928D+04    |proj g|=  7.38270D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    109      1     0     0   7.383D+00   3.099D+04
  F =   30992.812399692764

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.15678D+04    |proj g|=  1.11096D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    105      1     0     0   1.111D+01   3.157D+04
  F =   31567.796015751748

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.15645D+04    |proj g|=  1.11109D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.09901D+04    |proj g|=  7.38042D+00

At iterate  100    f=  3.30095D+04    |proj g|=  6.40468D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   6.405D+00   3.301D+04
  F =   33009.483212087900

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.30067D+04    |proj g|=  6.40487D+00

At iterate  100    f=  3.31285D+04    |proj g|=  1.18847D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   1.188D+01   3.313D+04
  F =   33128.459170482820

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.31258D+04    |proj g|=  1.18797D+01

At iterate   50    f=  3.22709D+04    |proj g|=  2.02002D+01

At iterate   50    f=  3.09805D+04    |proj g|=  5.53911D+00

At iterate   50    f=  3.15491D+04    |proj g|=  8.27267D+00

At iterate   50    f=  3.29966D+04    |proj g|=  1.00619D+01

At iterate   50    f=  3.31086D+04    |proj g|=  6.37572D+00

At iterate  100    f=  3.22532D+04    |proj g|=  1.12258D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.123D+01   3.225D+04
  F =   32253.171229828164

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.22528D+04    |proj g|=  1.12254D+01

At iterate  100    f=  3.15356D+04    |proj g|=  1.73320D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    105      1     0     0   1.733D+01   3.154D+04
  F =   31535.570385863415

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.09676D+04    |proj g|=  1.96023D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.960D+01   3.097D+04
  F =   30967.570517159849

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.15351D+04    |proj g|=  1.73315D+01
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.09672D+04    |proj g|=  1.96027D+01

At iterate  100    f=  3.29838D+04    |proj g|=  4.59068D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   4.591D+00   3.298D+04
  F =   32983.762119559331

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.29834D+04    |proj g|=  4.59065D+00

At iterate  100    f=  3.30958D+04    |proj g|=  4.64356D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   4.644D+01   3.310D+04
  F =   33095.848676494395

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =         5200     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  3.30955D+04    |proj g|=  4.64353D+01

At iterate   50    f=  3.22430D+04    |proj g|=  6.39587D+00

At iterate   50    f=  3.15266D+04    |proj g|=  1.91596D+01

At iterate   50    f=  3.09581D+04    |proj g|=  4.43404D+00

At iterate   50    f=  3.29741D+04    |proj g|=  1.85344D+01

At iterate   50    f=  3.30852D+04    |proj g|=  8.04640D+00

At iterate  100    f=  3.22340D+04    |proj g|=  7.42192D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   7.422D+00   3.223D+04
  F =   32234.027868082881

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.15146D+04    |proj g|=  6.80253D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    107      1     0     0   6.803D+00   3.151D+04
  F =   31514.581161989288

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.09528D+04    |proj g|=  6.39301D+00

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    110      1     0     0   6.393D+00   3.095D+04
  F =   30952.786412086960

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.29657D+04    |proj g|=  1.48928D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    108      1     0     0   1.489D+01   3.297D+04
  F =   32965.669129274691

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT

At iterate  100    f=  3.30737D+04    |proj g|=  1.99126D+01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
 5200    100    106      1     0     0   1.991D+01   3.307D+04
  F =   33073.655496377971

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT
</pre></div></div>
</div>
</section>
<section id="Prediction">
<h2>Prediction<a class="headerlink" href="#Prediction" title="Permalink to this heading">¶</a></h2>
<p>Use fitted model to predict celltypes in adata_pred (adata_test_corrected in our case). In the case of logistic regression, the threshold specifies the probability that needs to be reached to annotate a cell type or will be annotated as “unknown” if not reached. The threshold should be set to 0 or left out for SVM.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span>
    <span class="n">classifier</span><span class="o">=</span><span class="n">classifier</span><span class="p">,</span>
    <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">,</span>
    <span class="n">adata_pred</span><span class="o">=</span><span class="n">adata_test_corrected</span><span class="p">,</span>
    <span class="n">adata_orig</span><span class="o">=</span><span class="n">adata_test_orig</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>The prediction was added in a new column called ‘auto_annot’.</p>
<p>If in addition to the most likely class you would like to have all class probabilities returned use the following function. (This is only a sensible choice if using logistic regression.)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_pred_prob</span><span class="p">(</span>
    <span class="n">classifier</span><span class="o">=</span><span class="n">classifier</span><span class="p">,</span>
    <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">,</span>
    <span class="n">adata_pred</span><span class="o">=</span><span class="n">adata_test_corrected</span><span class="p">,</span>
    <span class="n">adata_orig</span><span class="o">=</span><span class="n">adata_test_orig</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="Output">
<h2>Output<a class="headerlink" href="#Output" title="Permalink to this heading">¶</a></h2>
<p>The adata object that includes the predicted cell type annotation can be written out as h5ad file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;/tmp/adata_predicted.h5ad&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>If the testing dataset included already a cell type annotation, a report can be generated and written, which includes metrics, confusion matrices and comparative umap plots. The report creates standardised overview figures, for a more detailed control of figure layout other besca and scanpy figure functions should be used.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
besca.tl.auto_annot.report(...) is deprecated( besca &gt; 2.3); please use besca.tl.report(...)
WARNING: saving figure to file figures/umap.ondata_auto_annot_tutorial.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_1.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_1.png" style="width: 1382px; height: 420px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_dblabel.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_3.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_3.png" style="width: 948px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_auto_annot.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_5.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_5.png" style="width: 1491px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_leiden.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_7.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_7.png" style="width: 469px; height: 460px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="admonition warning">
<p>Data type cannot be displayed: application/vnd.plotly.v1+json</p>
</div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_9.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_9.png" style="width: 955px; height: 635px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_45_10.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_45_10.png" style="width: 952px; height: 635px;" />
</div>
</div>
<p>If we change our threshold, we can vary the amount of cells labelled as unknown, depending on the confidence required from the predictions. Let’s try 0.7</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span>
    <span class="n">classifier</span><span class="o">=</span><span class="n">classifier</span><span class="p">,</span>
    <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">,</span>
    <span class="n">adata_pred</span><span class="o">=</span><span class="n">adata_test_corrected</span><span class="p">,</span>
    <span class="n">adata_orig</span><span class="o">=</span><span class="n">adata_test_orig</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>We observe that many more cells, where the assignment is not unambiguous are now labelled as unknown.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
besca.tl.auto_annot.report(...) is deprecated( besca &gt; 2.3); please use besca.tl.report(...)
WARNING: saving figure to file figures/umap.ondata_auto_annot_tutorial.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_1.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_1.png" style="width: 1382px; height: 420px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_dblabel.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_3.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_3.png" style="width: 948px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_auto_annot.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_5.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_5.png" style="width: 1491px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_leiden.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_7.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_7.png" style="width: 469px; height: 460px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="admonition warning">
<p>Data type cannot be displayed: application/vnd.plotly.v1+json</p>
</div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_9.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_9.png" style="width: 955px; height: 655px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_49_10.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_49_10.png" style="width: 952px; height: 655px;" />
</div>
</div>
</section>
<section id="Let’s-try-another-classifier-(SVM)">
<h2>Let’s try another classifier (SVM)<a class="headerlink" href="#Let’s-try-another-classifier-(SVM)" title="Permalink to this heading">¶</a></h2>
<p>Especially when thresholds are not used, SVMs can be tried as alternative classifiers.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="s2">&quot;auto_annot_tutorial_svm&quot;</span>
<span class="n">celltype</span> <span class="o">=</span> <span class="s2">&quot;dblabel&quot;</span>  <span class="c1"># column name of celltype of interest</span>
<span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>  <span class="c1"># rbf or linear or sgd, rbf extremely slow cannot recommend, logistic_regression recommended, as you can get probabilities, random_forest is a fast but not very powerful option (in current implementation)</span>
<span class="n">merge</span> <span class="o">=</span> <span class="s2">&quot;scanorama&quot;</span>
<span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">genes_to_use</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span>
    <span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="o">=</span><span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="n">merge</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
merging with scanorama
using scanorama rn
Found 207 genes among all datasets
[[0.   0.7  0.48]
 [0.   0.   0.99]
 [0.   0.   0.  ]]
Processing datasets (1, 2)
Processing datasets (0, 1)
Processing datasets (0, 2)
integrating training set
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
    <span class="n">adata_train</span><span class="o">=</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="o">=</span><span class="n">celltype</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">14</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span>
    <span class="n">classifier</span><span class="o">=</span><span class="n">classifier</span><span class="p">,</span>
    <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">,</span>
    <span class="n">adata_pred</span><span class="o">=</span><span class="n">adata_test_corrected</span><span class="p">,</span>
    <span class="n">adata_orig</span><span class="o">=</span><span class="n">adata_test_orig</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;/tmp/adata_predicted_svm.h5ad&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span>
                <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
besca.tl.auto_annot.report(...) is deprecated( besca &gt; 2.3); please use besca.tl.report(...)
WARNING: saving figure to file figures/umap.ondata_auto_annot_tutorial_svm.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_1.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_1.png" style="width: 1382px; height: 420px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_svm_dblabel.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_3.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_3.png" style="width: 948px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_svm_auto_annot.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_5.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_5.png" style="width: 1491px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_svm_leiden.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_7.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_7.png" style="width: 469px; height: 460px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="admonition warning">
<p>Data type cannot be displayed: application/vnd.plotly.v1+json</p>
</div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_9.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_9.png" style="width: 955px; height: 648px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_58_10.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_58_10.png" style="width: 952px; height: 648px;" />
</div>
</div>
</section>
<section id="Let’s-try-with-a-specified-gene-set">
<h2>Let’s try with a specified gene set<a class="headerlink" href="#Let’s-try-with-a-specified-gene-set" title="Permalink to this heading">¶</a></h2>
<p>Gene sets from GeMS or user specified genesets, can, if carefully chosen, significantly decrease computation time, withouth leading to a performance decrease.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="s2">&quot;auto_annot_tutorial_geneset&quot;</span>
<span class="n">celltype</span> <span class="o">=</span> <span class="s2">&quot;dblabel&quot;</span>  <span class="c1"># column name of celltype of interest</span>
<span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span>  <span class="c1"># rbf or linear or sgd, rbf extremely slow cannot recommend, logistic_regression recommended, as you can get probabilities, random_forest is a fast but not very powerful option (in current implementation)</span>
<span class="n">merge</span> <span class="o">=</span> <span class="s2">&quot;scanorama&quot;</span>
<span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">annotSigns</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">load_immune_signatures</span><span class="p">()</span>
<span class="n">allGenes</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">annotSigns</span><span class="p">:</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">annotSigns</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
<span class="n">genes_to_use</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">allGenes</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">annotSigns</span><span class="o">.</span><span class="n">values</span><span class="p">()))))</span>
<span class="n">display</span><span class="p">(</span><span class="n">genes_to_use</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;MLANA&#39;,
 &#39;LEF1&#39;,
 &#39;MMRN1&#39;,
 &#39;GZMA&#39;,
 &#39;MCM2&#39;,
 &#39;FGFBP2&#39;,
 &#39;KRT18&#39;,
 &#39;PCNA&#39;,
 &#39;TRDC&#39;,
 &#39;FABP4&#39;]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span>
    <span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="o">=</span><span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="n">merge</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
merging with scanorama
using scanorama rn
Found 77 genes among all datasets
[[0.   0.69 0.52]
 [0.   0.   0.98]
 [0.   0.   0.  ]]
Processing datasets (1, 2)
Processing datasets (0, 1)
Processing datasets (0, 2)
integrating training set
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span>
    <span class="n">classifier</span><span class="o">=</span><span class="n">classifier</span><span class="p">,</span>
    <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">,</span>
    <span class="n">adata_pred</span><span class="o">=</span><span class="n">adata_test_corrected</span><span class="p">,</span>
    <span class="n">adata_orig</span><span class="o">=</span><span class="n">adata_test_orig</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;/tmp/adata_predicted_svm.h5ad&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span>
<span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
besca.tl.auto_annot.report(...) is deprecated( besca &gt; 2.3); please use besca.tl.report(...)
WARNING: saving figure to file figures/umap.ondata_auto_annot_tutorial_geneset.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_1.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_1.png" style="width: 1382px; height: 420px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_geneset_dblabel.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_3.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_3.png" style="width: 948px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_geneset_auto_annot.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_5.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_5.png" style="width: 1491px; height: 415px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file figures/umap.auto_annot_tutorial_geneset_leiden.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_7.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_7.png" style="width: 469px; height: 460px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="admonition warning">
<p>Data type cannot be displayed: application/vnd.plotly.v1+json</p>
</div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_9.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_9.png" style="width: 955px; height: 641px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_auto_annot_tutorial_68_10.png" class="no-scaled-link" src="../_images/tutorials_auto_annot_tutorial_68_10.png" style="width: 952px; height: 641px;" />
</div>
</div>
</section>
<section id="Done!">
<h2>Done!<a class="headerlink" href="#Done!" title="Permalink to this heading">¶</a></h2>
</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../besca_standard_pipeline.html" class="btn btn-neutral float-right" title="besca standard pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="notebook3_batch_correction.html" class="btn btn-neutral float-left" title="notebook 3 - batch correction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, Roche Pharma Research and Early Development,     Pharmaceutical Sciences, Roche Innovation Center Basel, Basel, Switzerland.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>