

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Automated Annotation Workflow &mdash; BESCA 2.4 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery-dataframe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery-rendered-html.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../besca.html">besca</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials.html">tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Automated Annotation Workflow</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tutorials/auto_annot_tutorial-test_tmp.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Automated-Annotation-Workflow">
<h1>Automated Annotation Workflow<a class="headerlink" href="#Automated-Annotation-Workflow" title="Permalink to this headline">¶</a></h1>
<p>This workflow uses the auto_annot tools from besca to newly annotate a scRNAseq dataset based on one or more preannotated datasets. Ideally, these datasets come from a similar tissue and condition. If multiple training datasets are used, the performance is dependent on all dataset being annotated to the same resolution with broadly similar cell types.</p>
<p>We use supervised machine learning methods to annotate each individual cell utilizing methods like support vector machines (SVM) or logistic regression.</p>
<p>First, the traning dataset(s) and the testing dataset are loaded from h5ad files or made available as adata objects. Next, the training and testing datasets are corrected using scanorama, and the training datasets are then merged into one anndata object. Then, the classifier is trained utilizing the merged training data. Finally, the classifier is applied to the testing dataset to predict the cell types. If the testing dataset is already annotated (to test the algorithm), a report including
confusion matrices can be generated.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">besca</span> <span class="k">as</span> <span class="nn">bc</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
</div>
<div class="section" id="Parameter-specification">
<h2>Parameter specification<a class="headerlink" href="#Parameter-specification" title="Permalink to this headline">¶</a></h2>
<p>Give your analysis a name.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="s1">&#39;auto_annot_tutorial&#39;</span> <span class="c1"># The analysis name will be used to name the output files</span>
</pre></div>
</div>
</div>
<p>Specify column name of celltype annotation you want to train on.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">celltype</span> <span class="o">=</span><span class="s1">&#39;dblabel&#39;</span> <span class="c1"># This needs to be a column in the .obs of the training datasets (and test dataset if you want to generate a report)</span>
</pre></div>
</div>
</div>
<p>Choose a method:</p>
<ul class="simple">
<li><p>linear: Support Vector Machine with Linear Kernel</p></li>
<li><p>sgd: Support Vector Machine with Linear Kernel using Stochastic Gradient Descent</p></li>
<li><p>rbf: Support Vector Machine with radial basis function kernel. Very time intensive, use only on small datasets.</p></li>
<li><p>logistic_regression: Standard logistic classifier iwth multinomial loss.</p></li>
<li><p>logistic_regression_ovr: Logistic Regression with one versus rest classification.</p></li>
<li><p>logistic_regression_elastic: Logistic Regression with elastic loss, cross validates among multiple l1 ratios.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;logistic_regression&#39;</span>
</pre></div>
</div>
</div>
<p>Specify merge method. Needs to be either scanorama or naive.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">merge</span> <span class="o">=</span> <span class="s1">&#39;scanorama&#39;</span> <span class="c1"># We recommend to use scanorama here</span>
</pre></div>
</div>
</div>
<p>Decide if you want to use the raw format or highly variable genes. Raw increases computational time and does not necessarily improve predictions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># We recommend to use False here</span>
</pre></div>
</div>
</div>
<p>You can choose to only consider a subset of genes from a signature set or use all genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">genes_to_use</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span> <span class="c1"># We suggest to use all here, but the runtime is strongly improved if you select an appropriate gene set</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Data-loading">
<h2>Data loading<a class="headerlink" href="#Data-loading" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Read-in-all-training-sets-and-the-testing-set.">
<h3>Read in all training sets and the testing set.<a class="headerlink" href="#Read-in-all-training-sets-and-the-testing-set." title="Permalink to this headline">¶</a></h3>
<p>We will use a publicly available PBMC dataset, including ~3000 cells, as testing dataset. For the training dataset we use another PBMC dataset that is delivered with besca. The training datasets used are from:</p>
<p>Granja JM, Klemm S, McGinnis LM, et al. Single-cell multiomic analysis identifies regulatory programs in mixed-phenotype acute leukemia. Nat Biotechnol. 2019;37(12):1458-1465. doi:10.1038/s41587-019-0332-7</p>
<p>and</p>
<p>Kotliarov Y, Sparks R, Martins AJ, et al. Broad immune activation underlies shared set point signatures for vaccine responsiveness in healthy individuals and disease activity in patients with lupus. Nat Med. 2020;26(4):618-629. doi:10.1038/s41591-020-0769-8</p>
<p>A detailed analysis of the annotation performed below can be found in figure 3 of the publication on besca.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k_processed</span><span class="p">()</span> <span class="c1"># Dataset to be annotated (it&#39;s already annotated in this case for testing)</span>
<span class="n">adata_test_orig</span> <span class="o">=</span> <span class="n">adata_test</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># Make a copy of the testing data, which will be annotated in .obs, but not further modified</span>
<span class="n">adata_train_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">Granja2019_processed</span><span class="p">(),</span>
                    <span class="n">sc</span><span class="o">.</span><span class="n">rea</span><span class="p">]</span> <span class="c1"># List of annotated training datasets</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">OSError</span>                                   Traceback (most recent call last)
<span class="ansi-green-fg">~/Code/Besca/besca_dev/besca/datasets/_datasets.py</span> in <span class="ansi-cyan-fg">check_dl</span><span class="ansi-blue-fg">(filename, url)</span>
<span class="ansi-green-intense-fg ansi-bold">     27</span>     <span class="ansi-green-fg">try</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">---&gt; 28</span><span class="ansi-red-fg">         </span>adata <span class="ansi-blue-fg">=</span> read<span class="ansi-blue-fg">(</span>filename<span class="ansi-blue-fg">,</span> backup_url <span class="ansi-blue-fg">=</span> url<span class="ansi-blue-fg">,</span> cache<span class="ansi-blue-fg">=</span> <span class="ansi-green-fg">False</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     29</span>     <span class="ansi-green-fg">except</span> Exception<span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">~/.local/lib/python3.7/site-packages/scanpy/readwrite.py</span> in <span class="ansi-cyan-fg">read</span><span class="ansi-blue-fg">(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    121</span>             cache_compression<span class="ansi-blue-fg">=</span>cache_compression<span class="ansi-blue-fg">,</span>
<span class="ansi-green-fg">--&gt; 122</span><span class="ansi-red-fg">             </span><span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">    123</span>         )

<span class="ansi-green-fg">~/.local/lib/python3.7/site-packages/scanpy/readwrite.py</span> in <span class="ansi-cyan-fg">_read</span><span class="ansi-blue-fg">(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, suppress_cache_warning, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    699</span>         <span class="ansi-green-fg">if</span> sheet <span class="ansi-green-fg">is</span> <span class="ansi-green-fg">None</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 700</span><span class="ansi-red-fg">             </span><span class="ansi-green-fg">return</span> read_h5ad<span class="ansi-blue-fg">(</span>filename<span class="ansi-blue-fg">,</span> backed<span class="ansi-blue-fg">=</span>backed<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    701</span>         <span class="ansi-green-fg">else</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">~/.local/lib/python3.7/site-packages/anndata/_io/h5ad.py</span> in <span class="ansi-cyan-fg">read_h5ad</span><span class="ansi-blue-fg">(filename, backed, as_sparse, as_sparse_fmt, chunk_size)</span>
<span class="ansi-green-intense-fg ansi-bold">    399</span>
<span class="ansi-green-fg">--&gt; 400</span><span class="ansi-red-fg">     </span><span class="ansi-green-fg">with</span> h5py<span class="ansi-blue-fg">.</span>File<span class="ansi-blue-fg">(</span>filename<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#34;r&#34;</span><span class="ansi-blue-fg">)</span> <span class="ansi-green-fg">as</span> f<span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    401</span>         d <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">{</span><span class="ansi-blue-fg">}</span>

<span class="ansi-green-fg">~/.local/lib/python3.7/site-packages/h5py/_hl/files.py</span> in <span class="ansi-cyan-fg">__init__</span><span class="ansi-blue-fg">(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, **kwds)</span>
<span class="ansi-green-intense-fg ansi-bold">    407</span>                                fapl<span class="ansi-blue-fg">,</span> fcpl<span class="ansi-blue-fg">=</span>make_fcpl<span class="ansi-blue-fg">(</span>track_order<span class="ansi-blue-fg">=</span>track_order<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">,</span>
<span class="ansi-green-fg">--&gt; 408</span><span class="ansi-red-fg">                                swmr=swmr)
</span><span class="ansi-green-intense-fg ansi-bold">    409</span>

<span class="ansi-green-fg">~/.local/lib/python3.7/site-packages/h5py/_hl/files.py</span> in <span class="ansi-cyan-fg">make_fid</span><span class="ansi-blue-fg">(name, mode, userblock_size, fapl, fcpl, swmr)</span>
<span class="ansi-green-intense-fg ansi-bold">    172</span>             flags <span class="ansi-blue-fg">|=</span> h5f<span class="ansi-blue-fg">.</span>ACC_SWMR_READ
<span class="ansi-green-fg">--&gt; 173</span><span class="ansi-red-fg">         </span>fid <span class="ansi-blue-fg">=</span> h5f<span class="ansi-blue-fg">.</span>open<span class="ansi-blue-fg">(</span>name<span class="ansi-blue-fg">,</span> flags<span class="ansi-blue-fg">,</span> fapl<span class="ansi-blue-fg">=</span>fapl<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    174</span>     <span class="ansi-green-fg">elif</span> mode <span class="ansi-blue-fg">==</span> <span class="ansi-blue-fg">&#39;r+&#39;</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">h5py/_objects.pyx</span> in <span class="ansi-cyan-fg">h5py._objects.with_phil.wrapper</span><span class="ansi-blue-fg">()</span>

<span class="ansi-green-fg">h5py/_objects.pyx</span> in <span class="ansi-cyan-fg">h5py._objects.with_phil.wrapper</span><span class="ansi-blue-fg">()</span>

<span class="ansi-green-fg">h5py/h5f.pyx</span> in <span class="ansi-cyan-fg">h5py.h5f.open</span><span class="ansi-blue-fg">()</span>

<span class="ansi-red-fg">OSError</span>: Unable to open file (truncated file: eof = 744488960, sblock-&gt;base_addr = 0, stored_eof = 974992499)

During handling of the above exception, another exception occurred:

<span class="ansi-red-fg">URLError</span>                                  Traceback (most recent call last)
<span class="ansi-green-fg">&lt;ipython-input-9-9ebf32cc17fa&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">      1</span> adata_test <span class="ansi-blue-fg">=</span> bc<span class="ansi-blue-fg">.</span>datasets<span class="ansi-blue-fg">.</span>pbmc3k_processed<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span> <span class="ansi-red-fg"># Dataset to be annotated (it&#39;s already annotated in this case for testing)</span>
<span class="ansi-green-intense-fg ansi-bold">      2</span> adata_test_orig <span class="ansi-blue-fg">=</span> adata_test<span class="ansi-blue-fg">.</span>copy<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span> <span class="ansi-red-fg"># Make a copy of the testing data, which will be annotated in .obs, but not further modified</span>
<span class="ansi-green-fg">----&gt; 3</span><span class="ansi-red-fg"> </span>adata_train_list <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">[</span>bc<span class="ansi-blue-fg">.</span>datasets<span class="ansi-blue-fg">.</span>Granja2019_processed<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">,</span> bc<span class="ansi-blue-fg">.</span>datasets<span class="ansi-blue-fg">.</span>Kotliarov2020_processed<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">]</span> <span class="ansi-red-fg"># List of annotated training datasets</span>

<span class="ansi-green-fg">~/Code/Besca/besca_dev/besca/datasets/_datasets.py</span> in <span class="ansi-cyan-fg">Kotliarov2020_processed</span><span class="ansi-blue-fg">()</span>
<span class="ansi-green-intense-fg ansi-bold">    256</span>
<span class="ansi-green-intense-fg ansi-bold">    257</span>     filename <span class="ansi-blue-fg">=</span> pkg_resources<span class="ansi-blue-fg">.</span>resource_filename<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;besca&#39;</span><span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#39;datasets/data/Kotliarov2020_processed_citeseq_merged_annotated.h5ad&#39;</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-fg">--&gt; 258</span><span class="ansi-red-fg">     </span>adata <span class="ansi-blue-fg">=</span> check_dl<span class="ansi-blue-fg">(</span> filename<span class="ansi-blue-fg">,</span> url <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">&#39;https://zenodo.org/record/3938290/files/Kotliarov2020_processed_citeseq_merged_annotated.h5ad?download=1&#39;</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    259</span>     <span class="ansi-green-fg">return</span> adata
<span class="ansi-green-intense-fg ansi-bold">    260</span>

<span class="ansi-green-fg">~/Code/Besca/besca_dev/besca/datasets/_datasets.py</span> in <span class="ansi-cyan-fg">check_dl</span><span class="ansi-blue-fg">(filename, url)</span>
<span class="ansi-green-intense-fg ansi-bold">     28</span>         adata <span class="ansi-blue-fg">=</span> read<span class="ansi-blue-fg">(</span>filename<span class="ansi-blue-fg">,</span> backup_url <span class="ansi-blue-fg">=</span> url<span class="ansi-blue-fg">,</span> cache<span class="ansi-blue-fg">=</span> <span class="ansi-green-fg">False</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     29</span>     <span class="ansi-green-fg">except</span> Exception<span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">---&gt; 30</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">raise</span> URLError<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">f&#39;\n\n\n {filename} could not be downloaded from {url}; \n Please download it manually and store it in your besca installation: besca/datasets/data/&#39;</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     31</span>     <span class="ansi-green-fg">return</span> adata
<span class="ansi-green-intense-fg ansi-bold">     32</span>

<span class="ansi-red-fg">URLError</span>: &lt;urlopen error


 /pstore/home/julienla/Code/Besca/besca_dev/besca/datasets/data/Kotliarov2020_processed_citeseq_merged_annotated.h5ad could not be downloaded from https://zenodo.org/record/3938290/files/Kotliarov2020_processed_citeseq_merged_annotated.h5ad?download=1;
 Please download it manually and store it in your besca installation: besca/datasets/data/&gt;
</pre></div></div>
</div>
<p>You can also load your own dataset from the file system. Specify folders where .h5ad files are found and their names. The datasets that are already annotated and should be used for training. If you only use one dataset please use list of one. These load functions are useful in particular when the datasets have been processed with the besca standard pipeline. Alternatively, datasets can be loaded in any fashion, if more suitable.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#test_dataset_path = &#39;/path/to/test/dataset/folder&#39;</span>
<span class="c1">#test_dataset = &#39;testdataset.h5ad&#39;</span>
<span class="c1">#</span>
<span class="c1">#train_dataset_paths = [&#39;/path/to/train/dataset/folder1&#39;, &#39;/path/to/train/dataset/folder2&#39;]</span>
<span class="c1">#train_datasets = [&#39;traindataset1.h5ad&#39;, &#39;traindataset2.h5ad&#39;]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#adata_train_list, adata_test, adata_test_orig = bc.tl.auto_annot.read_data(train_paths = train_dataset_paths, train_datasets= train_datasets, test_path=  test_dataset_path, test_dataset= test_dataset, use_raw = use_raw)</span>
</pre></div>
</div>
</div>
<p>One training dataset’s label column is not named “dblabel”. We need to create a column so that all datasets used contain a column that are named according to the previously specified variable celltype and that contain the labels.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;dblabel&#39;</span><span class="p">]</span><span class="o">=</span> <span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype3&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>In this case the testing dataset is already annotated to demonstrate the methodology. All datasets adhere to the same naming convention.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_test</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dblabel</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Correct-datasets-(e.g. using-scanorama)-and-merge-training-datasets">
<h2>Correct datasets (e.g. using scanorama) and merge training datasets<a class="headerlink" href="#Correct-datasets-(e.g. using-scanorama)-and-merge-training-datasets" title="Permalink to this headline">¶</a></h2>
<p>This function merges training datasets, removes unwanted genes, and if scanorama is used corrects for datasets.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span><span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span> <span class="o">=</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span> <span class="o">=</span> <span class="n">merge</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Train-the-classifier">
<h2>Train the classifier<a class="headerlink" href="#Train-the-classifier" title="Permalink to this headline">¶</a></h2>
<p>The returned scaler is fitted on the training dataset (to zero mean and scaled to unit variance). The scaling will then be applied to the counts in the testing dataset and then the classifier is applied to the scaled testing dataset (see next step, adata_predict()). This function will run multiple jobs in parallel if if logistic regression was specified as method.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Prediction">
<h2>Prediction<a class="headerlink" href="#Prediction" title="Permalink to this headline">¶</a></h2>
<p>Use fitted model to predict celltypes in adata_pred (adata_test_corrected in our case). In the case of logistic regression, the threshold specifies the probability that needs to be reached to annotate a cell type or will be annotated as “unknown” if not reached. The threshold should be set to 0 or left out for SVM.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">scaler</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The prediction was added in a new column called ‘auto_annot’.</p>
<p>If in addition to the most likely class you would like to have all class probabilities returned use the following function. (This is only a sensible choice if using logistic regression.)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_pred_prob</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">scaler</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Output">
<h2>Output<a class="headerlink" href="#Output" title="Permalink to this headline">¶</a></h2>
<p>The adata object that includes the predicted cell type annotation can be written out as h5ad file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;/tmp/adata_predicted.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>If the testing dataset included already a cell type annotation, a report can be generated and written, which includes metrics, confusion matrices and comparative umap plots. The report creates standardised overview figures, for a more detailed control of figure layout other besca and scanpy figure functions should be used.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>If we change our threshold, we can vary the amount of cells labelled as unknown, depending on the confidence required from the predictions. Let’s try 0.7</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">scaler</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We observe that many more cells, where the assignment is not unambiguous are now labelled as unknown.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Let’s-try-another-classifier-(SVM)">
<h2>Let’s try another classifier (SVM)<a class="headerlink" href="#Let’s-try-another-classifier-(SVM)" title="Permalink to this headline">¶</a></h2>
<p>Especially when thresholds are not used, SVMs can be tried as alternative classifiers.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="s1">&#39;auto_annot_tutorial_svm&#39;</span>
<span class="n">celltype</span> <span class="o">=</span><span class="s1">&#39;dblabel&#39;</span> <span class="c1">#column name of celltype of interest</span>
<span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;linear&#39;</span>  <span class="c1"># rbf or linear or sgd, rbf extremely slow cannot recommend, logistic_regression recommended, as you can get probabilities, random_forest is a fast but not very powerful option (in current implementation)</span>
<span class="n">merge</span> <span class="o">=</span> <span class="s1">&#39;scanorama&#39;</span>
<span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">genes_to_use</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span><span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span> <span class="o">=</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span> <span class="o">=</span> <span class="n">merge</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">scaler</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;/tmp/adata_predicted_svm.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Let’s-try-with-a-specified-gene-set">
<h2>Let’s try with a specified gene set<a class="headerlink" href="#Let’s-try-with-a-specified-gene-set" title="Permalink to this headline">¶</a></h2>
<p>Gene sets from GeMS or user specified genesets, can, if carefully chosen, significantly decrease computation time, withouth leading to a performance decrease.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">analysis_name</span> <span class="o">=</span> <span class="s1">&#39;auto_annot_tutorial_geneset&#39;</span>
<span class="n">celltype</span> <span class="o">=</span><span class="s1">&#39;dblabel&#39;</span> <span class="c1">#column name of celltype of interest</span>
<span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;linear&#39;</span>  <span class="c1"># rbf or linear or sgd, rbf extremely slow cannot recommend, logistic_regression recommended, as you can get probabilities, random_forest is a fast but not very powerful option (in current implementation)</span>
<span class="n">merge</span> <span class="o">=</span> <span class="s1">&#39;scanorama&#39;</span>
<span class="n">use_raw</span> <span class="o">=</span> <span class="kc">False</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">annotSigns</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">load_immune_signatures</span><span class="p">()</span>
<span class="n">allGenes</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">annotSigns</span><span class="p">:</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">annotSigns</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
<span class="n">genes_to_use</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">allGenes</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">annotSigns</span><span class="o">.</span><span class="n">values</span><span class="p">()))))</span>
<span class="n">display</span><span class="p">(</span><span class="n">genes_to_use</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_test</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">(</span><span class="n">adata_train_list</span><span class="p">,</span> <span class="n">adata_test</span><span class="p">,</span> <span class="n">genes_to_use</span> <span class="o">=</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">merge</span> <span class="o">=</span> <span class="n">merge</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">celltype</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">scaler</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span> <span class="o">=</span> <span class="n">adata_test_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_predicted</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;/tmp/adata_predicted_svm.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">report</span><span class="p">(</span><span class="n">adata_predicted</span><span class="p">,</span> <span class="n">celltype</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="n">merge</span><span class="p">,</span> <span class="n">use_raw</span><span class="p">,</span> <span class="n">genes_to_use</span><span class="p">,</span> <span class="n">clustering</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Done!">
<h2>Done!<a class="headerlink" href="#Done!" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, BEDA.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>