

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>notebook 2 - celltype annotation and beyond &mdash; BESCA 2.3 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery-dataframe.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="notebook 3 - batch correction" href="notebook3_batch_correction.html" />
    <link rel="prev" title="notebook 1 - introduction and data processing" href="notebook1_data_processing_pbmc3k.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../besca.html">besca</a></li>
<li class="toctree-l1"><a class="reference internal" href="../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="notebook1_data_processing_pbmc3k.html">notebook 1 - introduction and data processing</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">notebook 2 - celltype annotation and beyond</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Dataset">Dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Celltype-annotation">Celltype annotation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Manual-cell-annotation-based-on-the-expression-of-known-marker-genes">Manual cell annotation based on the expression of known marker genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Hierarchical-signature-enrichment-based-approach-for-cell-type-annotation-for-Hematopoeitic-cells">Hierarchical signature-enrichment based approach for cell type annotation for Hematopoeitic cells</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Signature-loading-and-scoring">Signature loading and scoring</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Reclustering-Lymphocytes">Reclustering Lymphocytes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#unbiased-marker-genes-for-each-cluster">unbiased marker genes for each cluster</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Celltype-Quantification">Celltype Quantification</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebook3_batch_correction.html">notebook 3 - batch correction</a></li>
<li class="toctree-l2"><a class="reference internal" href="auto_annot_tutorial.html">Automated Annotation Workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#single-cell-sequencing-general-tutorials">single cell sequencing general tutorials</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#single-cell-auto-annot-tutorial-for-cell-type-annotation">single cell auto_annot tutorial for cell type annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#bescape-cell-deconvolution-tutorial">Bescape: cell deconvolution tutorial</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../tutorials.html">tutorials</a> &raquo;</li>
        
      <li>notebook 2 - celltype annotation and beyond</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/tutorials/notebook2_celltype_annotation_pbmc3k.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="notebook-2---celltype-annotation-and-beyond">
<h1>notebook 2 - celltype annotation and beyond<a class="headerlink" href="#notebook-2---celltype-annotation-and-beyond" title="Permalink to this headline">¶</a></h1>
<p>This notebook will introduce you to the process of celltype annotation and give you a brief outlook of some of the analysis you can do with single-cell data in scanpy &amp; besca.</p>
<p>Initial visualization and hand annotation is performed here to show the possible hand-annotation and how Besca is now proposing an automated annotation process, including for mix-cluster. This procedure is done in a more straighforward manner on the same dataset in the notebook available there:</p>
<p><a class="reference external" href="https://github.com/bedapub/besca_publication_results/blob/master/hematopoietic/pbmc3k/celltype_annotation_besca_Rc.ipynb">https://github.com/bedapub/besca_publication_results/blob/master/hematopoietic/pbmc3k/celltype_annotation_besca_Rc.ipynb</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#import necessary python packages</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span> <span class="c1">#software suite of tools for single-cell analysis in python</span>
<span class="kn">import</span> <span class="nn">besca</span> <span class="k">as</span> <span class="nn">bc</span> <span class="c1">#internal BEDA package for single cell analysis</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy</span>

<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span> <span class="c1">#output an overview of the software versions that are loaded</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.6.0 anndata==0.7.4 umap==0.4.6 numpy==1.19.2 scipy==1.5.2 pandas==1.1.3 scikit-learn==0.23.2 statsmodels==0.12.0 python-igraph==0.8.2 leidenalg==0.8.2
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Will only print error.</span>
<span class="c1"># Warning (eg genes not found) are ignored.</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="n">task</span> <span class="o">=</span> <span class="s2">&quot;&lt;style&gt;div.task { background-color: #ffc299; border-color: #ff944d; border-left: 7px solid #ff944d; padding: 1em;} &lt;/style&gt;&quot;</span>
<span class="n">HTML</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style>div.task { background-color: #ffc299; border-color: #ff944d; border-left: 7px solid #ff944d; padding: 1em;} </style></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">tag</span> <span class="o">=</span> <span class="s2">&quot;&lt;style&gt;div.tag { background-color: #99ccff; border-color: #1a8cff; border-left: 7px solid #1a8cff; padding: 1em;} &lt;/style&gt;&quot;</span>
<span class="n">HTML</span><span class="p">(</span><span class="n">tag</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style>div.tag { background-color: #99ccff; border-color: #1a8cff; border-left: 7px solid #1a8cff; padding: 1em;} </style></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">FAIR</span> <span class="o">=</span> <span class="s2">&quot;&lt;style&gt;div.fair { background-color: #d2f7ec; border-color: #d2f7ec; border-left: 7px solid #2fbc94; padding: 1em;} &lt;/style&gt;&quot;</span>
<span class="n">HTML</span><span class="p">(</span><span class="n">FAIR</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style>div.fair { background-color: #d2f7ec; border-color: #d2f7ec; border-left: 7px solid #2fbc94; padding: 1em;} </style></div>
</div>
<div class="section" id="Dataset">
<h2>Dataset<a class="headerlink" href="#Dataset" title="Permalink to this headline">¶</a></h2>
<p>Here we will reload our previously completely processed dataset.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#adata = sc.read(&#39;/tmp/adata_pbmc_FRESH_processed.h5ad&#39;)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k_filtered</span><span class="p">()</span>


</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2504 × 1719
    obs: &#39;CELL&#39;, &#39;percent_mito&#39;, &#39;experiment&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;leiden&#39;
    var: &#39;ENSEMBL&#39;, &#39;SYMBOL&#39;, &#39;n_cells&#39;, &#39;total_counts&#39;, &#39;frac_reads&#39;
    uns: &#39;experiment_colors&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Celltype-annotation">
<h2>Celltype annotation<a class="headerlink" href="#Celltype-annotation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Manual-cell-annotation-based-on-the-expression-of-known-marker-genes">
<h3>Manual cell annotation based on the expression of known marker genes<a class="headerlink" href="#Manual-cell-annotation-based-on-the-expression-of-known-marker-genes" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#define genes (in future you will be able to access genesets from a central database GEMS)</span>
<span class="n">b_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79A&#39;</span><span class="p">,</span> <span class="s1">&#39;MS4A1&#39;</span><span class="p">]</span>
<span class="n">t_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD3E&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3G&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3D&#39;</span><span class="p">]</span>
<span class="n">t_cell_subsets</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD4&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8B&#39;</span><span class="p">]</span>
<span class="n">naive_t_cell</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;SELL&#39;</span><span class="p">,</span> <span class="s1">&#39;CCR7&#39;</span><span class="p">,</span> <span class="s1">&#39;IL7R&#39;</span><span class="p">]</span>
<span class="n">myeloid_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s1">&#39;S100A9&#39;</span><span class="p">,</span> <span class="s1">&#39;CST3&#39;</span><span class="p">]</span>
<span class="n">monocytes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FCGR3A&#39;</span><span class="p">,</span> <span class="s1">&#39;FCGR3B&#39;</span><span class="p">,</span> <span class="s1">&#39;CD14&#39;</span><span class="p">]</span> <span class="c1">#FCGR3A/B = CD16</span>
<span class="n">dendritic_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FCER1A&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGAM&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGAX&#39;</span><span class="p">]</span> <span class="c1">#ITGAM = CD11b #ITGAX= CD11c</span>
<span class="n">NK_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;NCAM1&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3G&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="tag"><p>We always use the HGNC symbols of genes as the main annotation type in the AnnData object since they are the most practical to work with. In addition, the ensemble IDs of all genes is saved in adata.var according to our FAIR standard and always written out since Ensembl IDs stay very robust over time.</p>
<p>Based on the dataset the marker genes that are expressed may vary strongly.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
<span class="c1">#visualize the gene expression as an overlay of the umap</span>
<span class="c1">#(this way you can visually identify the clusters with a high expression))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">b_cells</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">t_cells</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">t_cell_subsets</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">naive_t_cell</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">NK_cells</span><span class="p">,</span>  <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span><span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">myeloid_cells</span><span class="p">,</span>  <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span><span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">monocytes</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">dendritic_cells</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_0.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_1.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_1.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_2.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_2.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_3.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_3.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_4.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_4.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_5.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_5.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_6.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_6.png" style="width: 1083px; height: 338px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_7.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_14_7.png" style="width: 1100px; height: 338px;" />
</div>
</div>
<div class="tag"><p>Per default scanpy plots the gene expression values saved in adata.raw (this means log1p(cp10k)). This is the reason that we can visualize all of these genes even if they are not contained in the highly variable genes.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_16_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_16_0.png" style="width: 376px; height: 334px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span> <span class="n">t_cell_subsets</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="s1">&#39;count&#39;</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_17_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_17_0.png" style="width: 971px; height: 340px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#write down new cluster names (important order needs to be equivalent to above)</span>
<span class="c1">#will label the cell types:</span>

<span class="n">new_cluster_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD14+ monocyte&#39;</span><span class="p">,</span> <span class="c1">#0</span>
                     <span class="s1">&#39;B-cell&#39;</span><span class="p">,</span> <span class="c1">#1</span>
                     <span class="s1">&#39;Mix T-cell&#39;</span><span class="p">,</span> <span class="c1">#2</span>
                     <span class="s1">&#39;CD8 T-cell&#39;</span><span class="p">,</span> <span class="c1">#3</span>
                     <span class="s1">&#39;Mix T-cell&#39;</span><span class="p">,</span> <span class="c1">#4</span>
                     <span class="s1">&#39;Mix T-cell&#39;</span><span class="p">,</span> <span class="c1">#5</span>
                     <span class="s1">&#39;CD8 T-cell&#39;</span><span class="p">,</span> <span class="c1">#6</span>
                     <span class="s1">&#39;FCGR3A+ monocyte&#39;</span><span class="p">,</span> <span class="c1">#7</span>
                     <span class="s1">&#39;CD8 T-cell&#39;</span><span class="p">,</span> <span class="c1">#8</span>
                     <span class="s1">&#39;NK cells&#39;</span><span class="p">,</span> <span class="c1">#9</span>
                     <span class="s1">&#39;Mix T-cell&#39;</span><span class="p">,</span> <span class="c1">#10</span>
                     <span class="s1">&#39;pDC&#39;</span><span class="p">]</span> <span class="c1">#1</span>

<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_cells_clustering</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">clustering_label</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">new_annotation_label</span><span class="o">=</span><span class="s1">&#39;manual_annotation&#39;</span><span class="p">,</span> <span class="n">new_cluster_labels</span><span class="o">=</span><span class="n">new_cluster_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="tag"><p>There are also more advanced methods of cell annotation where mixed clusters (i.e. clusters that contain more than one celltype) can be reclustered seperately and then annoated. Please refer to besca’s documentation for more information.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;manual_annotation&#39;</span><span class="p">],</span> <span class="n">wspace</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_20_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_20_0.png" style="width: 898px; height: 338px;" />
</div>
</div>
<div class="tag"><p>Several clusters have been combined into one celltype. Here only a very rudimentary celltype annotation was performed. Depending on the dataset and the type of analysis this can become far more complex and detailled.</p>
<p>We propose to use a an automated annotation methods to decipher the cluster in an <strong>un-bias</strong> way</p>
<p>This steps requires to expore initial labelling. This steps within besca will export:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">+</span> <span class="n">frac_pos</span><span class="p">,</span> <span class="n">computing</span> <span class="k">for</span> <span class="n">each</span> <span class="n">gene</span> <span class="ow">and</span> <span class="n">cluster</span> <span class="n">the</span> <span class="n">fraction</span> <span class="n">of</span> <span class="n">cells</span> <span class="n">that</span> <span class="n">show</span> <span class="n">a</span> <span class="n">gene</span> <span class="n">expression</span><span class="o">.</span>
<span class="o">+</span> <span class="n">average</span><span class="o">.</span><span class="n">gct</span><span class="p">,</span> <span class="n">computing</span> <span class="k">for</span> <span class="n">each</span> <span class="n">gene</span> <span class="n">the</span> <span class="n">average</span> <span class="n">expression</span> <span class="n">per</span> <span class="n">cluster</span>
</pre></div>
</div>
</div><div class="fair"><p>write out required annotation
<code class="docutils literal notranslate"><span class="pre">bc.export.labeling(adata=adata,</span> <span class="pre">outpath=FILEPATH_TO_FOLDER,</span> <span class="pre">column</span>&#160; <span class="pre">=</span> <span class="pre">'leiden')</span></code></p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">bc</span><span class="o">.</span><span class="n">export</span><span class="o">.</span><span class="n">labeling</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">outpath</span><span class="o">=</span><span class="s1">&#39;dataset_export/&#39;</span><span class="p">,</span> <span class="n">column</span>  <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
mapping of cells to  leiden exported successfully to cell2labels.tsv
average.gct exported successfully to file
fract_pos.gct exported successfully to file
</pre></div></div>
</div>
</div>
<div class="section" id="Hierarchical-signature-enrichment-based-approach-for-cell-type-annotation-for-Hematopoeitic-cells">
<h3>Hierarchical signature-enrichment based approach for cell type annotation for Hematopoeitic cells<a class="headerlink" href="#Hierarchical-signature-enrichment-based-approach-for-cell-type-annotation-for-Hematopoeitic-cells" title="Permalink to this headline">¶</a></h3>
<p>Besca offers a hierarchical signature-enrichment based approach for cell type annotation. The initial hierarchy can be visualized below</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pkg_resources</span>

<span class="n">config_file</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;besca&#39;</span><span class="p">,</span> <span class="s1">&#39;datasets/genesets/CellNames_scseqCMs6_config.tsv&#39;</span><span class="p">)</span>
<span class="n">plt</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">nomenclature_network</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_25_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_25_0.png" style="width: 610px; height: 610px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Signature-loading-and-scoring">
<h2>Signature loading and scoring<a class="headerlink" href="#Signature-loading-and-scoring" title="Permalink to this headline">¶</a></h2>
<p>We load signature directly importing a gmt file. In case of connection to a GemS database, it is possible to use the GeMs signature loader.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sigconfig</span><span class="p">,</span><span class="n">levsk</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">read_annotconfig</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gmt_file_anno</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;besca&#39;</span><span class="p">,</span> <span class="s1">&#39;datasets/genesets/CellNames_scseqCMs6_sigs.gmt&#39;</span><span class="p">)</span>


<span class="n">mymarkers</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">read_GMT_sign</span><span class="p">(</span><span class="n">gmt_file_anno</span><span class="p">,</span><span class="n">directed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mymarkers</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">filter_siggenes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">mymarkers</span><span class="p">)</span>


<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">combined_signature_score</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gmt_file_anno</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1">#</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">scores</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;scanpy&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span> <span class="n">scores</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_30_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_30_0.png" style="width: 819px; height: 5059px;" />
</div>
</div>
<div class="tag"><p>Signature can be observed in detailled to assess manually the cell types to which assign the different clusters. However Besca offers to use a signature-based hierarchical annotation to build on know relations (see network above). This annotation is based on the mann whitney scoring of the positive fraction of cell and signatures provided. Additionally, this signature-based annotation procedures will then use a consistent labelling of the dataset and cell-types, facilitating cross-study
comparisons in the futur.</p>
</div><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="n">mymarkers</span><span class="p">[</span><span class="s1">&#39;Ubi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;B2M&#39;</span><span class="p">,</span><span class="s1">&#39;ACTB&#39;</span><span class="p">,</span> <span class="s1">&#39;GAPDH&#39;</span><span class="p">]</span> <span class="c1">### used for cutoff adjustment to individual dataset, can be modified</span>

<span class="n">f</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;dataset_export/fract_pos.gct&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">df</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">score_mw</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">mymarkers</span><span class="p">)</span>

<span class="n">myc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Ubi&#39;</span><span class="p">,:]</span><span class="o">*</span><span class="mf">0.5</span><span class="p">)</span> <span class="c1">### Set a cutoff based on Ubi and scale with values from config file</span>
</pre></div>
</div>
</div>
<p>One can observe the resulting log10(pvalues) obtained</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">7</span><span class="p">]</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>6</th>
      <th>0</th>
      <th>7</th>
      <th>9</th>
      <th>8</th>
      <th>11</th>
      <th>1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Bcell</th>
      <td>10.5977</td>
      <td>4.33419</td>
      <td>7.80649</td>
      <td>1.49797</td>
      <td>2.68703</td>
      <td>7.80042</td>
      <td>69.0433</td>
    </tr>
    <tr>
      <th>Fibroblast</th>
      <td>0.789622</td>
      <td>0.239568</td>
      <td>1.19473</td>
      <td>0.942461</td>
      <td>0.285982</td>
      <td>0.290622</td>
      <td>0.155626</td>
    </tr>
    <tr>
      <th>Endothelial</th>
      <td>1.71538</td>
      <td>0.904445</td>
      <td>9.77757</td>
      <td>0.814998</td>
      <td>0.461955</td>
      <td>6.4328</td>
      <td>0.116507</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Cluster attribution based on cutoff</span>
<span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;Ubi&#39;</span><span class="p">)</span>
<span class="n">sigscores</span><span class="o">=</span><span class="p">{}</span>
<span class="k">for</span> <span class="n">mysig</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
    <span class="n">sigscores</span><span class="p">[</span><span class="n">mysig</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">getset</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">mysig</span><span class="p">,</span><span class="n">sigconfig</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mysig</span><span class="p">,</span><span class="s1">&#39;Cutoff&#39;</span><span class="p">]</span><span class="o">*</span><span class="n">myc</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sigconfig</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;CD8Tcell&quot;</span><span class="p">,</span><span class="s2">&quot;Cutoff&quot;</span><span class="p">]</span><span class="o">=</span><span class="mf">1.2</span>
<span class="c1"># RECOMPUTING SIG SCORE WITH NEW CUTOFF</span>
<span class="n">sigscores</span><span class="o">=</span><span class="p">{}</span>
<span class="k">for</span> <span class="n">mysig</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
    <span class="n">sigscores</span><span class="p">[</span><span class="n">mysig</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">getset</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">mysig</span><span class="p">,</span><span class="n">sigconfig</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mysig</span><span class="p">,</span><span class="s1">&#39;Cutoff&#39;</span><span class="p">]</span><span class="o">*</span><span class="n">myc</span><span class="p">)</span>


<span class="n">cnames</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">make_anno</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">sigscores</span><span class="p">,</span><span class="n">sigconfig</span><span class="p">,</span><span class="n">levsk</span><span class="p">)</span>
<span class="n">cnames</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>celltype0</th>
      <th>celltype1</th>
      <th>celltype2</th>
      <th>celltype3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD8Tcell</td>
      <td>NaiCD8Tcell</td>
    </tr>
    <tr>
      <th>0</th>
      <td>Hematopoietic</td>
      <td>Myeloid</td>
      <td>ClassMonocyte</td>
      <td>ClassMonocyte</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Hematopoietic</td>
      <td>Myeloid</td>
      <td>NClassMonocyte</td>
      <td>NClassMonocyte</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Hematopoietic</td>
      <td>NKcell</td>
      <td>CD56dimNK</td>
      <td>CD56dimNK</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD8Tcell</td>
      <td>NaiCD8Tcell</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Hematopoietic</td>
      <td>Myeloid</td>
      <td>cDC</td>
      <td>cDC2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Hematopoietic</td>
      <td>Blymphocyte</td>
      <td>Bcell</td>
      <td>NaiBcell</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD8Tcell</td>
      <td>CytotoxCD8Tcell</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD4Tcell</td>
      <td>NaiCD4Tcell</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD4Tcell</td>
      <td>EMCD4Tcell</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD4Tcell</td>
      <td>NaiCD4Tcell</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Hematopoietic</td>
      <td>Tcell</td>
      <td>CD4Tcell</td>
      <td>NaiCD4Tcell</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">label_files</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">resource_filename</span><span class="p">(</span><span class="s1">&#39;besca&#39;</span><span class="p">,</span> <span class="s1">&#39;/datasets/nomenclature/CellTypes_v1.tsv&#39;</span><span class="p">)</span>

<span class="n">cnamesDBlabel</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">obtain_dblabel</span><span class="p">(</span> <span class="n">label_files</span><span class="p">,</span> <span class="n">cnames</span> <span class="p">)</span>
<span class="n">cnamesDBlabel</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>celltype0</th>
      <th>celltype1</th>
      <th>celltype2</th>
      <th>celltype3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD8-positive, alpha-beta T cell</td>
      <td>naive thymus-derived CD8-positive, alpha-beta ...</td>
    </tr>
    <tr>
      <th>0</th>
      <td>hematopoietic cell</td>
      <td>myeloid leukocyte</td>
      <td>classical monocyte</td>
      <td>classical monocyte</td>
    </tr>
    <tr>
      <th>7</th>
      <td>hematopoietic cell</td>
      <td>myeloid leukocyte</td>
      <td>non-classical monocyte</td>
      <td>non-classical monocyte</td>
    </tr>
    <tr>
      <th>9</th>
      <td>hematopoietic cell</td>
      <td>natural killer cell</td>
      <td>cytotoxic CD56-dim natural killer cell</td>
      <td>cytotoxic CD56-dim natural killer cell</td>
    </tr>
    <tr>
      <th>8</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD8-positive, alpha-beta T cell</td>
      <td>naive thymus-derived CD8-positive, alpha-beta ...</td>
    </tr>
    <tr>
      <th>11</th>
      <td>hematopoietic cell</td>
      <td>myeloid leukocyte</td>
      <td>myeloid dendritic cell</td>
      <td>CD1c-positive myeloid dendritic cell</td>
    </tr>
    <tr>
      <th>1</th>
      <td>hematopoietic cell</td>
      <td>lymphocyte of B lineage</td>
      <td>B cell</td>
      <td>naive B cell</td>
    </tr>
    <tr>
      <th>3</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD8-positive, alpha-beta T cell</td>
      <td>CD8-positive, alpha-beta cytotoxic T cell</td>
    </tr>
    <tr>
      <th>10</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD4-positive, alpha-beta T cell</td>
      <td>naive thymus-derived CD4-positive, alpha-beta ...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD4-positive, alpha-beta T cell</td>
      <td>effector memory CD4-positive, alpha-beta T cell</td>
    </tr>
    <tr>
      <th>4</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD4-positive, alpha-beta T cell</td>
      <td>naive thymus-derived CD4-positive, alpha-beta ...</td>
    </tr>
    <tr>
      <th>5</th>
      <td>hematopoietic cell</td>
      <td>T cell</td>
      <td>CD4-positive, alpha-beta T cell</td>
      <td>naive thymus-derived CD4-positive, alpha-beta ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype0&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype0&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype1&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype1&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype2&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype3&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype3&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype1&#39;</span><span class="p">,</span>
                        <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8B&#39;</span><span class="p">,</span> <span class="c1"># CD8 T cell markers</span>
                        <span class="s1">&#39;CD4&#39;</span> <span class="p">,</span>    <span class="s1">&#39;GPR183&#39;</span><span class="p">,</span> <span class="s1">&#39;CMTM8&#39;</span> <span class="c1"># CD4 t cell markers</span>
                       <span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="c1">#,&#39;celltype3&#39;</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_39_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_39_0.png" style="width: 575px; height: 709px;" />
</div>
</div>
<div class="tag"><p>Here we believe some clusters of T cells are still mixed and that a more refined annotation can be done if recluster the lymphocytes. This procedure is applied below</p>
</div></div>
<div class="section" id="Reclustering-Lymphocytes">
<h2>Reclustering Lymphocytes<a class="headerlink" href="#Reclustering-Lymphocytes" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_original&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_rc</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">recluster</span> <span class="p">(</span> <span class="n">adata</span><span class="p">,</span> <span class="n">celltype_label</span> <span class="o">=</span> <span class="s1">&#39;celltype1&#39;</span><span class="p">,</span>
                               <span class="n">celltype</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;T cell&#39;</span><span class="p">,</span>  <span class="s1">&#39;natural killer cell&#39;</span><span class="p">),</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">1.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
In total 1409 highly variable genes selected within cluster
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span> <span class="n">adata_rc</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span>
                              <span class="c1"># Some markers</span>
                              <span class="s1">&#39;CD3G&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4&#39;</span><span class="p">,</span> <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;GNLY&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_44_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_44_0.png" style="width: 804px; height: 361px;" />
</div>
</div>
<div class="tag"><p>Exporting the additional labelling allows to regenerate frac_pos required file for the annotation function.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_rc</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">st</span><span class="o">.</span><span class="n">additional_labeling</span><span class="p">(</span><span class="n">adata_rc</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;Leiden_Reclustering&#39;</span><span class="p">,</span> <span class="s1">&#39;Leiden Reclustering on Lymphocytes&#39;</span><span class="p">,</span> <span class="s1">&#39;BescaTutorial&#39;</span><span class="p">,</span> <span class="s1">&#39;dataset_export&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
rank genes per label calculated using method wilcoxon.
mapping of cells to  leiden exported successfully to cell2labels.tsv
average.gct exported successfully to file
fract_pos.gct exported successfully to file
labelinfo.tsv successfully written out
dataset_export/labelings/Leiden_Reclustering/WilxRank.gct written out
dataset_export/labelings/Leiden_Reclustering/WilxRank.pvalues.gct written out
dataset_export/labelings/Leiden_Reclustering/WilxRank.logFC.gct written out
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mymarkers</span><span class="p">[</span><span class="s1">&#39;Ubi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;B2M&#39;</span><span class="p">,</span><span class="s1">&#39;ACTB&#39;</span><span class="p">,</span> <span class="s1">&#39;GAPDH&#39;</span><span class="p">]</span>
<span class="n">clusters</span> <span class="o">=</span> <span class="s1">&#39;Leiden_Reclustering&#39;</span>
<span class="n">f</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;dataset_export&#39;</span> <span class="o">+</span> <span class="s2">&quot;/labelings/&quot;</span><span class="o">+</span><span class="n">clusters</span><span class="o">+</span><span class="s2">&quot;/fract_pos.gct&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">df</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">score_mw</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">mymarkers</span><span class="p">)</span>
<span class="n">myc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;Ubi&#39;</span><span class="p">,:]</span><span class="o">*</span><span class="mf">0.5</span><span class="p">)</span> <span class="c1">### Set a cutoff based on Ubi and scale with values from config file</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># RECOMPUTING SIG SCORE WITH NEW CUTOFF</span>
<span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;Ubi&#39;</span><span class="p">)</span>
<span class="n">sigscores</span><span class="o">=</span><span class="p">{}</span>
<span class="k">for</span> <span class="n">mysig</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
    <span class="n">sigscores</span><span class="p">[</span><span class="n">mysig</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">getset</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">mysig</span><span class="p">,</span><span class="n">sigconfig</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">mysig</span><span class="p">,</span><span class="s1">&#39;Cutoff&#39;</span><span class="p">]</span><span class="o">*</span><span class="n">myc</span><span class="p">)</span>
    <span class="c1">#sigscores[mysig]=bc.tl.sig.getset(df,mysig,10)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">toExclude</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span>  <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">levsk</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;Tcell&#39;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;NKcell&#39;</span><span class="p">]</span>

<span class="n">cnames</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">make_anno</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">sigscores</span><span class="p">,</span><span class="n">sigconfig</span><span class="p">,</span><span class="n">levsk</span><span class="p">,</span>  <span class="n">toexclude</span><span class="o">=</span> <span class="n">toExclude</span><span class="p">)</span>

<span class="n">cnamesDBlabel</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">obtain_dblabel</span><span class="p">(</span><span class="n">label_files</span><span class="p">,</span> <span class="n">cnames</span> <span class="p">)</span>


<span class="n">adata_rc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype0&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata_rc</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype0&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="n">adata_rc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype2&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata_rc</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="n">adata_rc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype3&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">sig</span><span class="o">.</span><span class="n">add_anno</span><span class="p">(</span><span class="n">adata_rc</span><span class="p">,</span><span class="n">cnamesDBlabel</span><span class="p">,</span><span class="s1">&#39;celltype3&#39;</span><span class="p">,</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_rc</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;celltype3&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype2&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;CD3G&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4&#39;</span><span class="p">,</span> <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;GNLY&#39;</span><span class="p">,</span> <span class="c1">#, &#39;S100A&#39;</span>
                           <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;SELL&#39;</span>
                          <span class="p">],</span>
           <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_50_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_50_0.png" style="width: 454px; height: 1753px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">names_2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">names_3</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span> <span class="n">cnames</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="p">:</span>
    <span class="c1"># Orderigng lexo. order</span>
    <span class="n">names_2</span> <span class="o">+=</span> <span class="p">[</span><span class="n">cnamesDBlabel</span><span class="p">[</span><span class="s1">&#39;celltype2&#39;</span><span class="p">][</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]]</span>
    <span class="n">names_3</span> <span class="o">+=</span> <span class="p">[</span><span class="n">cnamesDBlabel</span><span class="p">[</span><span class="s1">&#39;celltype3&#39;</span><span class="p">][</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">annotate_new_cellnames</span><span class="p">(</span> <span class="n">adata</span><span class="p">,</span> <span class="n">adata_rc</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">names_2</span><span class="p">,</span> <span class="n">new_label</span><span class="o">=</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>

<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">annotate_new_cellnames</span><span class="p">(</span> <span class="n">adata</span><span class="p">,</span> <span class="n">adata_rc</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">names_3</span><span class="p">,</span> <span class="n">new_label</span><span class="o">=</span><span class="s1">&#39;celltype3&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
NOTE: overwriting labels for the selected cells saved in adata.obs.celltype2 with the new labels
NOTE: overwriting labels for the selected cells saved in adata.obs.celltype3 with the new labels
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;celltype2&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;celltype3&#39;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span> <span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_53_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_53_0.png" style="width: 454px; height: 361px;" />
</div>
</div>
<div class="tag"><p>Here we write out the celltype annotation so that we can use reload it in later analysis steps. There is a corresponding besca function for reloading the annotation.</p>
</div><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#save AnnData Object for comparision purposes</span>
<span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;dataset_export/data_pbmc_processed_annotated.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="unbiased-marker-genes-for-each-cluster">
<h2>unbiased marker genes for each cluster<a class="headerlink" href="#unbiased-marker-genes-for-each-cluster" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#perform differential gene expression between each cluster and all other cells using scanpy</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;celltype3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#visualize the top 3 marker genes as a dot plot</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_dotplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_58_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_58_0.png" style="width: 555px; height: 481px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_matrixplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_59_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_59_0.png" style="width: 555px; height: 481px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_stacked_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_60_0.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_60_0.png" style="width: 558px; height: 481px;" />
</div>
</div>
</div>
<div class="section" id="Celltype-Quantification">
<h2>Celltype Quantification<a class="headerlink" href="#Celltype-Quantification" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#counts per celltype</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">count_occurrence</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">add_percentage</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">count_variable</span><span class="o">=</span><span class="s1">&#39;celltype3&#39;</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Counts</th>
      <th>Percentage</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>classical monocyte</th>
      <td>438</td>
      <td>17.49</td>
    </tr>
    <tr>
      <th>naive thymus-derived CD4-positive, alpha-beta T cell</th>
      <td>414</td>
      <td>16.53</td>
    </tr>
    <tr>
      <th>naive thymus-derived CD8-positive, alpha-beta T cell</th>
      <td>391</td>
      <td>15.62</td>
    </tr>
    <tr>
      <th>naive B cell</th>
      <td>317</td>
      <td>12.66</td>
    </tr>
    <tr>
      <th>CD8-positive, alpha-beta cytotoxic T cell</th>
      <td>271</td>
      <td>10.82</td>
    </tr>
    <tr>
      <th>central memory CD4-positive, alpha-beta T cell</th>
      <td>267</td>
      <td>10.66</td>
    </tr>
    <tr>
      <th>non-classical monocyte</th>
      <td>159</td>
      <td>6.35</td>
    </tr>
    <tr>
      <th>IL7R-max CD8-positive, alpha-beta cytotoxic T cell</th>
      <td>118</td>
      <td>4.71</td>
    </tr>
    <tr>
      <th>CD8-positive, alpha-beta T cell</th>
      <td>100</td>
      <td>3.99</td>
    </tr>
    <tr>
      <th>CD1c-positive myeloid dendritic cell</th>
      <td>29</td>
      <td>1.16</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#generate a basic bar plot of the table above</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>

<span class="c1">#visualize distribution before logarithmization</span>
<span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;Counts&#39;</span> <span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

<span class="c1">#visualize distribution after logarithmization</span>
<span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;Percentage&#39;</span> <span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
 [Text(0, 0, &#39;classical monocyte&#39;),
  Text(1, 0, &#39;naive thymus-derived CD4-positive, alpha-beta T cell&#39;),
  Text(2, 0, &#39;naive thymus-derived CD8-positive, alpha-beta T cell&#39;),
  Text(3, 0, &#39;naive B cell&#39;),
  Text(4, 0, &#39;CD8-positive, alpha-beta cytotoxic T cell&#39;),
  Text(5, 0, &#39;central memory CD4-positive, alpha-beta T cell&#39;),
  Text(6, 0, &#39;non-classical monocyte&#39;),
  Text(7, 0, &#39;IL7R-max CD8-positive, alpha-beta cytotoxic T cell&#39;),
  Text(8, 0, &#39;CD8-positive, alpha-beta T cell&#39;),
  Text(9, 0, &#39;CD1c-positive myeloid dendritic cell&#39;)])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_63_1.png" class="no-scaled-link" src="../_images/tutorials_notebook2_celltype_annotation_pbmc3k_63_1.png" style="width: 630px; height: 502px;" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="notebook3_batch_correction.html" class="btn btn-neutral float-right" title="notebook 3 - batch correction" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="notebook1_data_processing_pbmc3k.html" class="btn btn-neutral float-left" title="notebook 1 - introduction and data processing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, BEDA

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>