

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>annotate celltypes &mdash; BESCA 2.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="tutorials" href="../../tutorials.html" />
    <link rel="prev" title="reclustering on specific louvain clusters" href="../tools/plot_reclustering_function.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../besca.html">besca</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">code examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../index.html#plotting-examples">Plotting examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../index.html#preprocessing">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../index.html#tools">Tools</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../index.html#workflows">Workflows</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">annotate celltypes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#visualization-of-marker-genes">visualization of marker genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cluster-level-celltype-annotation">cluster level celltype annotation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reclustering-on-mixed-cell-clusters">reclustering on mixed cell clusters</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials.html">tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Code Examples</a> &raquo;</li>
        
      <li>annotate celltypes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/auto_examples/workflows/plot_celltype_annotation.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="sphx-glr-download-link-note admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Click <a class="reference internal" href="#sphx-glr-download-auto-examples-workflows-plot-celltype-annotation-py"><span class="std std-ref">here</span></a> to download the full example code</p>
</div>
<div class="sphx-glr-example-title section" id="annotate-celltypes">
<span id="sphx-glr-auto-examples-workflows-plot-celltype-annotation-py"></span><h1>annotate celltypes<a class="headerlink" href="#annotate-celltypes" title="Permalink to this headline">¶</a></h1>
<p>An example workflow using the PBMC3k dataset included with besca illustrating how to annotate celltypes based on louvain clusters.
This workflow begins with a preprocessed and filtered dataset on which a louvain clustering was already performed.
Please refer to other tutorials on how to perform these steps.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#load libraries</span>
<span class="kn">import</span> <span class="nn">besca</span> <span class="kn">as</span> <span class="nn">bc</span>
<span class="kn">import</span> <span class="nn">scanpy.api</span> <span class="kn">as</span> <span class="nn">sc</span>

<span class="c1">#load preprocessed dataset (included in BESCA for demonstration purposes)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k_processed</span><span class="p">()</span>

<span class="c1">#need to drop celltype annotation stored in this dataset (only relevant for this tutorial)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">],</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>

<span class="c1">#visualize the louvain clusters</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;louvain&#39;</span><span class="p">])</span>
</pre></div>
</div>
<img alt="../../_images/sphx_glr_plot_celltype_annotation_001.png" class="sphx-glr-single-img" src="../../_images/sphx_glr_plot_celltype_annotation_001.png" />
<div class="section" id="visualization-of-marker-genes">
<h2>visualization of marker genes<a class="headerlink" href="#visualization-of-marker-genes" title="Permalink to this headline">¶</a></h2>
<p>Depending on the type of data you are analysing you will need to look at
different marker genes that are specific to the celltypes you would expect
to find in your dataset. In this case we are looking at PBMC cells and will
try to identify the main Immunecell subtypes: T-cells, B-cells, Monocytes, and
Dendritic cells.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#identification of T-cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD3E&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3G&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3D&#39;</span><span class="p">])</span>

<span class="c1">#identification of NK cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;NCAM1&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;GNLY&#39;</span><span class="p">])</span>

<span class="c1">#identification of B-cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79A&#39;</span><span class="p">])</span>

<span class="c1">#identification of myeloid cells/dendritic cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CST3&#39;</span><span class="p">,</span> <span class="s1">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s1">&#39;S100A9&#39;</span><span class="p">])</span>

<span class="c1">#identification of dendritic cells(FCERIA) and monocytes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FCER1A&#39;</span><span class="p">,</span><span class="s1">&#39;CD14&#39;</span><span class="p">,</span> <span class="s1">&#39;FCGR3A&#39;</span><span class="p">])</span>
</pre></div>
</div>
<ul class="sphx-glr-horizontal">
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_002.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_002.png" />
</li>
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_003.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_003.png" />
</li>
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_004.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_004.png" />
</li>
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_005.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_005.png" />
</li>
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_006.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_006.png" />
</li>
</ul>
</div>
<div class="section" id="cluster-level-celltype-annotation">
<h2>cluster level celltype annotation<a class="headerlink" href="#cluster-level-celltype-annotation" title="Permalink to this headline">¶</a></h2>
<p>Depending on how fine-grained your clustering is you will often come into the
the situation that a louvain cluster contains several cell-populations that
are clearly segregated based on the marker gene expression. If this happens you
can try and adjust the louvain resolution parameter to make the clustering more
fine-grained, but this will not always be successfull. Especially in cases where
your sample contains vastly different celltypes (e.g. an Immuncell dataset
containing B-cells and T-cells) it will be difficult to resolve T-cell subsets
since they are much more comparable to each other than e.g. a T-cell vs a B-cell.
In this case it often makes sense to make a highlevel cell-labeling and then perform
a second clustering on onoly the mixed cellclusters. This is the procedure that will
be demonstrated in the rest of this tutorial.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#define high-level celltype annotation</span>
<span class="n">new_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;mixed&quot;</span><span class="p">,</span> <span class="c1">#0</span>
              <span class="s2">&quot;mixed&quot;</span><span class="p">,</span> <span class="c1">#1</span>
              <span class="s2">&quot;CD14+ monocyte&quot;</span><span class="p">,</span> <span class="c1">#2</span>
              <span class="s2">&quot;mixed&quot;</span><span class="p">,</span> <span class="c1">#3</span>
              <span class="s2">&quot;B-cell&quot;</span><span class="p">,</span> <span class="c1">#4</span>
              <span class="s2">&quot;FCGR3A+ monocyte&quot;</span><span class="p">,</span> <span class="c1">#5</span>
              <span class="s2">&quot;mixed&quot;</span><span class="p">,</span> <span class="c1">#6</span>
              <span class="s2">&quot;pDC&quot;</span><span class="p">]</span> <span class="c1">#7</span>

<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate_cells_clustering</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">new_labels</span><span class="p">)</span>

<span class="c1">#visualize annotation</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">])</span>

<span class="c1">#preserve highlevel labels for future use if desired</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;high_level celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">celltype</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../../_images/sphx_glr_plot_celltype_annotation_007.png" class="sphx-glr-single-img" src="../../_images/sphx_glr_plot_celltype_annotation_007.png" />
</div>
<div class="section" id="reclustering-on-mixed-cell-clusters">
<h2>reclustering on mixed cell clusters<a class="headerlink" href="#reclustering-on-mixed-cell-clusters" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#perform reclustering on subset using besca function</span>
<span class="n">adata_subset</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cluster</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;6&#39;</span><span class="p">),</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">1.3</span><span class="p">)</span>

<span class="c1">#visualize important marker genes in reclustering</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_subset</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3G&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4&#39;</span><span class="p">,</span> <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;GNLY&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>

<span class="c1">#annotate celltypes based on the new louvain clusters</span>
<span class="n">new_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#0</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#1</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#2</span>
              <span class="s2">&quot;CD8 T-cell&quot;</span><span class="p">,</span> <span class="c1">#3</span>
              <span class="s2">&quot;NK cell&quot;</span><span class="p">,</span> <span class="c1">#4</span>
              <span class="s2">&quot;CD8 T-cell&quot;</span><span class="p">,</span> <span class="c1">#5</span>
              <span class="s2">&quot;CD8 T-cell&quot;</span><span class="p">,</span><span class="c1">#6</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#7</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#8</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">,</span> <span class="c1">#9</span>
              <span class="s2">&quot;CD4 T-cell&quot;</span><span class="p">]</span> <span class="c1">#10</span>

<span class="c1">#merge new celllabels back into the original adata object containing all cells</span>
<span class="c1">#Note: this will overwrite the labels contained in adata.obs.celltype! If you w</span>
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">annotate_new_cellnames</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_subset</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">new_labels</span><span class="p">)</span>

<span class="c1">#visualize finished celltype annotation</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">])</span>
</pre></div>
</div>
<ul class="sphx-glr-horizontal">
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_008.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_008.png" />
</li>
<li><img alt="../../_images/sphx_glr_plot_celltype_annotation_009.png" class="sphx-glr-multi-img first" src="../../_images/sphx_glr_plot_celltype_annotation_009.png" />
</li>
</ul>
<p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>In total 1471 highly variable genes selected within cluster
NOTE: overwriting labels for the selected cells saved in adata.obs.celltype with the new labels
</pre></div>
</div>
<p><strong>Total running time of the script:</strong> ( 0 minutes  15.606 seconds)</p>
<div class="sphx-glr-footer class sphx-glr-footer-example docutils container" id="sphx-glr-download-auto-examples-workflows-plot-celltype-annotation-py">
<div class="sphx-glr-download docutils container">
<a class="reference download internal" download="" href="../../_downloads/c8cee6f8919b3193f3584e88cb038fc7/plot_celltype_annotation.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_celltype_annotation.py</span></code></a></div>
<div class="sphx-glr-download docutils container">
<a class="reference download internal" download="" href="../../_downloads/cc4f2f43b0822807339bb0de25769a51/plot_celltype_annotation.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_celltype_annotation.ipynb</span></code></a></div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.readthedocs.io">Gallery generated by Sphinx-Gallery</a></p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../../tutorials.html" class="btn btn-neutral float-right" title="tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../tools/plot_reclustering_function.html" class="btn btn-neutral float-left" title="reclustering on specific louvain clusters" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, BEDA

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>