{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotating Cell Types\n",
    "\n",
    "This workbook can be run after the standard workflow.\n",
    "It shows how to use the  annotation function to annotate the dataset that was generated by the standard worflow to assign clusters to cell types.\n",
    "\n",
    "We also demonstrate signature-scoring functions and data exploration that can further facilitate annotation or the evaluation thereof.\n",
    "\n",
    "An alternative in the case a annotated training dataset already exists is to use the auto-annot module. Please refer to the corresponding tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import besca as bc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import sparse, io\n",
    "import os\n",
    "import time\n",
    "import logging\n",
    "import seaborn as sns\n",
    "sc.logging.print_header()\n",
    "\n",
    "# for standard processing, set verbosity to minimum\n",
    "sc.settings.verbosity = 0  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "version = '2.8'\n",
    "start0 = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#define standardized filepaths based on above input\n",
    "root_path = os.getcwd()\n",
    "bescapath_full = os.path.dirname(bc.__file__)\n",
    "bescapath = os.path.split(bescapath_full)[0]\n",
    "\n",
    "analysis_name = 'standard_workflow_besca2'\n",
    "species='mouse' ## or mouse for now\n",
    "conversion=None\n",
    "sigsuffix=''\n",
    "if species=='mouse': sigsuffix='.mouse'\n",
    "#analysis_name = 'standard_workflow_besca2_CLR' #use _CLR or _DSB for citeseq analysis\n",
    "annot_author = 'annot_author' ### replace with userid\n",
    "\n",
    "# Choose the clustering to base the annotation on.\n",
    "# leiden (RNA), citeseq (protein)\n",
    "clusters='leiden' \n",
    "\n",
    "# Which UMAP basis to use for visualization\n",
    "# umap or umap_citeseq\n",
    "umap_basis='umap'\n",
    "\n",
    "# Specify if Cite-seq data\n",
    "citeseq=False\n",
    "\n",
    "# Specify if signatures should be exported to gmtx\n",
    "export_sigs=False\n",
    "\n",
    "# Specify if you are connected to a mongoDB that you want to use to explore similarity with previous data\n",
    "mongodb_explore=False\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The chunk of code below is usefull if this is the initial installation of besca and that you are running this notebook as a test. It will download if needed the test datasets and export the labelling. \n",
    "This export is usually done at the end of the standard workflow. Exported files  are necessary for the annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_example_dataset = False\n",
    "if use_example_dataset:\n",
    "    analysis_name='pbmc3k_processed'\n",
    "    results_folder = os.path.split(os.getcwd())[0] + '/besca/datasets/data/'\n",
    "    clusters='leiden'\n",
    "    umap_basis='umap'\n",
    "    citeseq = False\n",
    "    # This line will either download, or load the datasets\n",
    "    adata = bc.datasets.pbmc3k_processed()\n",
    "    # This line exports the annotation for the annotation.\n",
    "    adata = bc.st.additional_labeling(adata, labeling_to_use= clusters, labeling_name = clusters, \n",
    "                                      labeling_description = 'Exporting a postori the labels for annotation',\n",
    "                                      labeling_author = 'Testing', \n",
    "                                      results_folder= results_folder)\n",
    "else:\n",
    "    if clusters == 'leiden':\n",
    "        results_folder = os.path.join(root_path, 'analyzed')\n",
    "        if citeseq:\n",
    "            results_folder = os.path.join(results_folder, analysis_name, 'citeseq' , 'citeseq') \n",
    "            adata = sc.read_h5ad(os.path.join(results_folder + '_merged' ,analysis_name + '.h5ad') ) \n",
    "        else:\n",
    "            results_folder = os.path.join(results_folder, analysis_name)\n",
    "            adata = sc.read_h5ad(os.path.join(results_folder, analysis_name + '.h5ad') )\n",
    "    else:\n",
    "        results_folder = os.path.join(root_path, 'analyzed', analysis_name, 'citeseq', clusters)\n",
    "        adata = sc.read_h5ad(os.path.join(results_folder + '_merged' ,analysis_name + '.h5ad') )\n",
    "        clusters='leiden_citeseq'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create export file and folder names\n",
    "results_file = os.path.join(results_folder, analysis_name + '.annotated.h5ad')\n",
    "figdir=os.path.join(root_path, 'analyzed', analysis_name+'/figures/')\n",
    "sc.settings.figdir = figdir\n",
    "if not os.path.exists(figdir):\n",
    "    os.makedirs(figdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata, color = [clusters], basis = umap_basis )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if citeseq:\n",
    "    # Expression pattern of the antibodies\n",
    "    # Can be usefull to refine the annotations, if the clusters are RNA-based\n",
    "    ab_names = adata.var[adata.var.feature_type=='Antibody Capture']\n",
    "    sc.pl.embedding(adata, basis = umap_basis, color=ab_names.index.tolist(), color_map = 'viridis')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore top marker genes per cluster "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEgenes=bc.tl.dge.get_de(adata,clusters,demethod='wilcoxon',topnr=5000, logfc=1,padj=0.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select only top genes (in order of p-val) for 2 clusters and plot expression per cluster\n",
    "### Sort according to FC for more specific genes\n",
    "tops=list(DEgenes['0'].sort_values('Log2FC',ascending=False)['Name'][0:20])+list(DEgenes['1'].sort_values('Log2FC',ascending=False)['Name'][0:20])\n",
    "sc.pl.dotplot(adata, var_names=tops,groupby=clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Explore similarity to other data in MongoDB (only works if connected to MongoDB with studies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if mongodb_explore:\n",
    "    import sys\n",
    "    sys.path.append(pathtomongoloader) #Pls adjust e.g. /bioinfo/scseq/mongoloader/\n",
    "    from scCorr import clusterCorrelation\n",
    "\n",
    "    fract_pos=results_folder+'/labelings/'+clusters+'/fract_pos.gct'\n",
    "    \n",
    "    ### Select clusters of interest e.g. 0|1 or \"ALL\" for all clusters; \n",
    "    ### Select studies of interest for correlating: dbstudies=studyID|analysisID|labelingID or dbstudies=ALL; species=mouse (or human) or species=ALL\n",
    "    \n",
    "    ### Select general studies first (use more stringent cutoff e.g. 0.3 to reduce correlations)\n",
    "    cordf=clusterCorrelation(fract_pos,  \"-clusters=1|5\", \"mouse\", \"-dbstudies=ALL\", \"-dbspecies=mouse\", 0.3)\n",
    "    cordf[0] # These are all the top hits\n",
    "    cordf[1].index=[x.split('   ')[0]+'_'+x.split('   ')[3] for x in list(cordf[1].index)] #Simplify labeling for heatmap\n",
    "    ### Plot top results as heatmap\n",
    "    sns.set_context(\"paper\", font_scale=1.1)     \n",
    "    sns.clustermap(cordf[1].loc[cordf[1].transpose().max()>=0.45,:], figsize=(4,10), annot=True,  cmap=\"YlGnBu\")\n",
    "    \n",
    "    \n",
    "    ### Second example - select a specific study\n",
    "    cordf=clusterCorrelation(fract_pos,  \"-clusters=0|1\", \"mouse\", \"-dbstudies=TabulaMuris_droplet|TM_droplet_all_besca2_7|celltype\", \"-dbspecies=mouse\", 0.1)\n",
    "    cordf[0] ### This is the top hit per cluster\n",
    "    cordf[1].index=[x.split('   ')[3] for x in list(cordf[1].index)] #Simplify labeling for heatmap\n",
    "    ### Plot top results as heatmap\n",
    "    sns.set_context(\"paper\", font_scale=1.1)     \n",
    "    sns.clustermap(cordf[1].loc[cordf[1].transpose().max()>=0.3,:], figsize=(5,6), annot=True,  cmap=\"YlGnBu\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore distribution of various cell populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# One can load besca-provided signatures using the function below\n",
    "signature_dict = bc.datasets.load_immune_signatures(refined=False)\n",
    "\n",
    "signature_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionaly it is possible to read an compute scanpy score using this function below.\n",
    "\n",
    "If the gmt file is composed of combined signature (UP and DN), a common score will be computed: \n",
    "$$Total\\_SCORE= Score_{UP} - Score_{DN}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gmt_file= bescapath + '/besca/datasets/genesets/Immune.gmt'\n",
    "\n",
    "if species=='mouse':\n",
    "    # Genes converted to mouse homologs\n",
    "    mousehuman_file = bescapath + '/besca/datasets/homologs/MGItoHGNC.csv'\n",
    "    mousehuman=pd.read_csv(mousehuman_file,sep='\\t',header='infer', encoding=\"unicode_escape\")\n",
    "    mousehuman.index=mousehuman['MGI']\n",
    "    conversion=pd.Series(data=mousehuman['HGNC'], index=mousehuman.index)\n",
    "\n",
    "bc.tl.sig.combined_signature_score(adata, gmt_file,\n",
    "                             UP_suffix='_UP', DN_suffix='_DN', method='scanpy',\n",
    "                             overwrite=False, verbose=False,\n",
    "                             use_raw=True, conversion=conversion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "scores = [x for x in adata.obs.columns if 'scanpy' in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata, basis = umap_basis, color= scores)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Signatures for specific sub-populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Provided with besca; change this for own gmt file\n",
    "gmt_file_anno= bescapath + '/besca/datasets/genesets/CellNames_scseqCMs6_sigs'+sigsuffix+'.gmt'\n",
    "bc.tl.sig.combined_signature_score(adata, gmt_file_anno) #optional conversion argument , conversion=conversion\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot all signatures containing \"scanpy\" in name\n",
    "scores = [x for x in adata.obs.columns if 'scanpy' in x]\n",
    "sc.pl.embedding(adata, basis = umap_basis, color= scores, color_map = 'viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## An extra set of signatures (less specific but informative) is also provided \n",
    "gmt_file_anno_extra= bescapath + '/besca/datasets/genesets/CellNames_scseqCMs6_Extrasigs'+sigsuffix+'.gmt'\n",
    "bc.tl.sig.combined_signature_score(adata, gmt_file_anno_extra) #optional conversion argument , conversion=conversion\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot all signatures containing \"_scv\" in name\n",
    "scores = [x for x in adata.obs.columns if '_scv' in x]\n",
    "sc.pl.embedding(adata, basis = umap_basis, color= scores, color_map = 'viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Plot only selected signatures\n",
    "sc.pl.embedding(adata, basis = umap_basis, color= ['score_Myeloid_scanpy','score_Bcell_scanpy','score_Tcell_scanpy','score_NKcell_scanpy'], color_map = 'viridis')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# Automated annotation\n",
    "\n",
    "A decision-tree-based annotation that reads signatures from a provided .gmt file and hierarchy as well as cutoffs and signature ordering from a configuration file and attributes each cell to a specific type according to signature enrichment. \n",
    "\n",
    "This is an aid to start ther annotation and annotation can then be further refined by adding further signatures or adjusting the configuration files. It was tested mainly on PBMCs and oncology (tumor biopsies) related samples.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading markers and signature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import repeat\n",
    "mymarkers = bc.tl.sig.read_GMT_sign(gmt_file_anno,directed=False)\n",
    "\n",
    "#optional conversion - if human-based signatures are read\n",
    "#if species=='mouse':\n",
    "#    for signature in mymarkers.keys():\n",
    "#        mymarkers[signature] = [i for i in map(bc.tl.sig._helper._to_geneid, repeat(conversion), mymarkers[signature]) if i is not None]\n",
    "\n",
    "mymarkers = bc.tl.sig.filter_siggenes(adata, mymarkers) ### remove genes not present in dataset or empty signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mymarkers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select Ubiquitously expressed genes for cutoff adjustment to invidual datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Cutoff in the configuration file will be calculated relative to the enrichment of this ubiquitous signature\n",
    "## For best performance, choose genes that are uniformly distributed across all clusters on a given dataset \n",
    "## Ideally, expression is moderate rather than high\n",
    "mymarkers['Ubi'] = [ 'ZNF207', 'HNRNPU','SNRPD3', 'SRRM1'] # alternatives: ['B2M','ACTB', 'HNRNPK'] \n",
    "if species=='mouse':\n",
    "    mymarkers['Ubi'] = ['B2m','Actb',  'Hnrnpk','Hnrnpu'] # alternative: 'Znf207'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Inspect gene expression for ubi genes, checking for uniform distribution across clusters\n",
    "sc.pl.embedding(adata, basis = umap_basis, color= mymarkers['Ubi'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, var_names= mymarkers['Ubi'], groupby=clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Inspect gene expression for an example signature\n",
    "sc.pl.embedding(adata, basis = umap_basis, color= mymarkers['Hematopoietic'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configuration of the annotation\n",
    "\n",
    "We read the configuration file, containing hierarchy, cutoff and signature priority information. \n",
    "A new version of this file should be created and maintained with each annotation. \n",
    "The included example is optimised for the annotation of the 6.6k PBMC dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "configfile=bescapath + '/besca/datasets/genesets/CellNames_scseqCMs6_config'+sigsuffix+'.tsv' ### replace this with your config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigconfig,levsk=bc.tl.sig.read_annotconfig(configfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional configuatation: The order of cells at different levels can be manually changed if needed\n",
    "#levsk[0]=['ColorectalCancer', 'Epithelial','Fibroblast','Endothelial','Erythrocyte','HematoStem',\n",
    "#          'Hepatocyte','MelMelanoma','Neural','Adipocyte','Hematopoietic','Schwann','Chondrocyte','Glial']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get an overview of the cell type hierarchy included in the configuration file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt=bc.pl.nomenclature_network(configfile, font_size=8)\n",
    "plt.savefig(figdir+\"Nomenclatureplot.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fract_pos was exported by BESCA in the standard worflow test, \n",
    "contains information of fraction positive cells per genes per cluster.\n",
    "\n",
    "We use these values as a basis for a wilcoxon test per signature per cluster. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Optional: For an alternative clustering to be used as annotation, one needs to export the corresponding fract_pos\n",
    "\n",
    "#adata = bc.st.clustering(adata, results_folder, myres=2, method = clusters) ## higher clustering resolution\n",
    "#bc.export.clustering(adata, outpath = os.path.join(results_folder, 'labelings', clusters+'_r2'), method = clusters)\n",
    "#bc.export.labeling_info(outpath=os.path.join(results_folder, 'labelings', clusters+'_r2'), description=clusters+' clustering with r=2', method='leiden')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=pd.read_csv(results_folder + '/labelings/' + clusters + '/fract_pos.gct' ,sep=\"\\t\",skiprows=2)\n",
    "df=bc.tl.sig.score_mw(f,mymarkers)\n",
    "\n",
    "### Set a cutoff based on Ubi and scale with values from config file\n",
    "### Change the factor 0.5 to systematically be more stringent (higher e.g. 1) or leniant (lower e.g. 0.25)\n",
    "myfactor=0.5\n",
    "if species=='mouse':\n",
    "    myfactor=0.3\n",
    "myc=np.median(df.loc['Ubi',:]*myfactor) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check the cutoff \n",
    "myc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check the ubiquitous signature score (should be as uniform as possible across clusters)\n",
    "df.loc[\"Ubi\",:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[0:3,0:7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### For an overview of highest scoring signatures, one can generate a heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.clustermap(df.loc[df.max(axis=1)>myc*1.5,:].astype(float),figsize=(8, 12))\n",
    "plt.savefig(figdir+\"SignatureHeatmap_all.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each signature, positive and negative clusters are determined. Only positive clusters are maintained. Cutoffs can be individualised based on the config file (scaling factor) and myc, which is determined based on ubiquitously expressed genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove ubiquitous signature for the scoring part\n",
    "df=df.drop('Ubi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional: Cutoffs can also be manually adjusted if needed \n",
    "# Always adjust from lowest to highest level and remember to check priorities as well (order)\n",
    "# sigconfig.loc['Epithelial','Cutoff']=1.5 ### Increase for being more stringent, decrease for more leniant\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cluster attribution based on cutoff (all clusters above cutoff will be attributed to a cell type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigscores={}\n",
    "for mysig in list(df.index):\n",
    "    sigscores[mysig]=bc.tl.sig.getset(df,mysig,sigconfig.loc[mysig,'Cutoff']*myc)\n",
    "    #sigscores[mysig]=bc.tl.sig.getset(df,mysig,10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One can inspect the cluster attribution per cell type in the signature list and adjust cutoffs as required. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigscores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check to a specific cell type \n",
    "sigscores['Hematopoietic']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now each cluster gets annotated, according to the distinct levels specified in the config file. \n",
    "Note that in case a cluster is positive for multiple identities, only the first one is taken, \n",
    "in the order specified in the \"Order\" column in the config file. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To check the given order, per levels, you can inspect levsk, and adjust above as needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "levsk"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell types that are not expected in the dataset or that are too fine-grained be explicity excluded from the annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "### For instance, if no erythrocytes or pancreatic cells are expected, they can be specified here\n",
    "### For instance, if plasma cells should not be subclassified, they can can be specified here\n",
    "toexclude=['Erythrocyte','AlphaPancreatic', 'BetaPancreatic', \n",
    "           'DeltaPancreatic','IgGPlasma','IgAPlasma','IgMPlasma','PancreaticDuctal','Cholangiocyte']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain cluster assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnames=bc.tl.sig.make_anno(df,sigscores,sigconfig,levsk, toexclude=toexclude)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now obtained per each cluster cell type attribution at distinct levels. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export the used annotation parameters, for future reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "bc.tl.sig.export_annotconfig(sigconfig, levsk, results_folder, analysis_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using dblabel convention\n",
    "\n",
    "Only short names were used in the signature naming convention in this case. \n",
    "One can easity tranform this to EFO terms if preferred, a conversion table comes with besca. \n",
    "\n",
    "This nomenclature is quite extended, and the function \n",
    "**obtain_dblabel** can perform the conversion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Transform these short forms to dblabel - EFO standard nomenclature\n",
    "cnamesDBlabel = bc.tl.sig.obtain_dblabel(bescapath+'/besca/datasets/nomenclature/CellTypes_v1.tsv', cnames )\n",
    "cnamesDBlabel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, one can add the new labels to adata.obs as annotation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['celltype0']=bc.tl.sig.add_anno(adata,cnamesDBlabel,'celltype0',clusters)\n",
    "adata.obs['celltype1']=bc.tl.sig.add_anno(adata,cnamesDBlabel,'celltype1',clusters)\n",
    "adata.obs['celltype2']=bc.tl.sig.add_anno(adata,cnamesDBlabel,'celltype2',clusters)\n",
    "adata.obs['celltype3']=bc.tl.sig.add_anno(adata,cnamesDBlabel,'celltype3',clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inspect the labels on the umap from lowest to highest resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color=['celltype1'], basis = umap_basis) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color=['celltype2'], basis = umap_basis) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color=['celltype3'], basis = umap_basis) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper functions for additional checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Which classification has cluster 7? \n",
    "bc.tl.sig.match_cluster(adata,'leiden','7','celltype3',0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### What cluster corresponds to CD1c-positive myeloid dendritic cell? \n",
    "bc.tl.sig.match_cluster(adata,'celltype3','CD1c-positive myeloid dendritic cell','leiden',0.3) ## lowered cutoff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual refinement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In some cases, the annotation does not produce the optimal result. Manual adjustments can be made, \n",
    "by replacing cell type names or by manually labeling clusters. Note that the second option is not \n",
    "stable across reruns if any adjustments are made to the clustering. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Example 1: Classical monocytes contains both classical and non-classical monocytes, adjust to monocytes\n",
    "#adata.obs['celltype2']=list(adata.obs['celltype2'].replace('classical monocyte', 'monocyte'))\n",
    "#adata.obs['celltype3']=list(adata.obs['celltype3'].replace('classical monocyte', 'monocyte'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Example 2: non-classical monocytes need to be added to a specific cluster\n",
    "#adata.obs['celltype2']=adata.obs['celltype2'].cat.add_categories('non-classical monocyte')\n",
    "#adata.obs['celltype3']=adata.obs['celltype3'].cat.add_categories('non-classical monocyte')\n",
    "#adata.obs.loc[adata.obs[clusters].isin(['4']),'celltype2']='non-classical monocyte'\n",
    "#adata.obs.loc[adata.obs[clusters].isin(['4']),'celltype3']='non-classical monocyte'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Example 3: if some epithelial cells are not recognised as cancer cells\n",
    "#adata.obs.loc[adata.obs['celltype3']=='epithelial cell','celltype0']='colorectal cancer cell'\n",
    "#adata.obs.loc[adata.obs['celltype3']=='epithelial cell','celltype1']='colorectal cancer cell'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Make sure to remove obsolete categories\n",
    "adata.obs['celltype3']=adata.obs['celltype3'].cat.remove_unused_categories()\n",
    "adata.obs['celltype2']=adata.obs['celltype2'].cat.remove_unused_categories()\n",
    "adata.obs['celltype1']=adata.obs['celltype1'].cat.remove_unused_categories()\n",
    "adata.obs['celltype0']=adata.obs['celltype0'].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reclustering sub-clusters \n",
    "\n",
    "Sometimes, single clusters can contain a mix of cell types. For example, for the PBMC3K data example, the lymphocyte clusters are mixed. In this case, one can try to increase clustering resolution (as above) or recluster specifically on those clusters.\n",
    "Below we show an example.\n",
    "\n",
    "The main steps are:\n",
    "+ Saving previous clustering and annotation for comparison purpose (advised)\n",
    "+ Recluster \n",
    "+ Export the new labelling (see function additional_labeling)\n",
    "+ Read the new labelling information including fract_pos files\n",
    "+ Recompute signatures/markers values\n",
    "+ Reannotate\n",
    "+ Convert annotation to dblabel\n",
    "+ Export all for the data subset to the larger adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "recluster = False\n",
    "\n",
    "if use_example_dataset:\n",
    "    recluster = True\n",
    "    celltype_label = 'celltype2_original'\n",
    "    to_recluster =  ('CD8-positive, alpha-beta T cell','CD4-positive, alpha-beta T cell',\n",
    "                                         'cytotoxic CD56-dim natural killer cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if recluster:\n",
    "    # Save previous clustering obtained for comparision\n",
    "    adata.obs['leiden_original'] = adata.obs['leiden'].copy()\n",
    "    adata.obs['celltype2_original']  = adata.obs['celltype2'].copy() \n",
    "\n",
    "    # Calling reclustering\n",
    "    adata_rc = bc.tl.rc.recluster ( adata, celltype_label = celltype_label, \n",
    "                               celltype=to_recluster, resolution=1.3)\n",
    "\n",
    "    # Leiden reclustering have to be exported to use the annotation function \n",
    "    cluster_renamed = 'Leiden_Reclustering'\n",
    "    adata_rc = bc.st.additional_labeling(adata_rc, 'leiden', cluster_renamed, 'Leiden Reclustering on Lymphocytes', 'author', results_folder)\n",
    "   \n",
    "    # Reading additional labelling\n",
    "    f=pd.read_csv(results_folder + \"/labelings/\"+cluster_renamed+\"/fract_pos.gct\",sep=\"\\t\",skiprows=2)\n",
    "    df=bc.tl.sig.score_mw(f,mymarkers)\n",
    "    myc=np.median(df.loc['Ubi',:]*0.5) ### Set a cutoff based on Ubi and scale with values from config file\n",
    "\n",
    "    # Recomputing the signature score with the new cutoff\n",
    "    df=df.drop('Ubi')\n",
    "    sigscores={}\n",
    "    for mysig in list(df.index):\n",
    "        sigscores[mysig]=bc.tl.sig.getset(df,mysig,sigconfig.loc[mysig,'Cutoff']*myc)\n",
    "\n",
    "    # Adjust cutoffs if needed\n",
    "    # sigconfig.loc['Epithelial','Cutoff']=1.5\n",
    "    \n",
    "    cnames=bc.tl.sig.make_anno(df,sigscores,sigconfig,levsk, toexclude=toexclude)\n",
    "    cnamesDBlabel = bc.tl.sig.obtain_dblabel(bescapath+'/besca/datasets/nomenclature/CellTypes_v1.tsv', cnames )\n",
    "    \n",
    "    adata_rc.obs['celltype0']=bc.tl.sig.add_anno(adata_rc,cnamesDBlabel,'celltype0','leiden')\n",
    "    adata_rc.obs['celltype2']=bc.tl.sig.add_anno(adata_rc,cnamesDBlabel,'celltype2','leiden')\n",
    "    adata_rc.obs['celltype3']=bc.tl.sig.add_anno(adata_rc,cnamesDBlabel,'celltype3','leiden')\n",
    "    # Lex order needed.\n",
    "    names_2 = []\n",
    "    names_3 = []\n",
    "    for i in range( cnames.shape[0]) :\n",
    "        names_2 += [cnames['celltype2'][str(i)]]\n",
    "        names_3 += [cnames['celltype3'][str(i)]]\n",
    "    \n",
    "    bc.tl.rc.annotate_new_cellnames( adata, adata_rc, names = names_2, new_label='celltype2', method = 'leiden')\n",
    "\n",
    "    bc.tl.rc.annotate_new_cellnames( adata, adata_rc, names = names_3, new_label='celltype3', method = 'leiden')\n",
    "    \n",
    "    sc.pl.umap(adata,color=['celltype2', 'celltype2_original',\n",
    "                       'celltype3'], ncols=1) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save annotation result and export labelling\n",
    "\n",
    "Chosen labels can also be exported as a new folder in labelings/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Attribute the cell annotation level of choice (typically the highest resolution one, if estimated to be reliable) to dblabel\n",
    "adata.obs['dblabel']=adata.obs['celltype3']\n",
    "\n",
    "### Export file for future reference\n",
    "adata.write(results_file)\n",
    "#adata=sc.read(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save labelling celltype1\n",
    "adata = bc.st.additional_labeling(adata, 'celltype1', 'celltype1', 'Major cell types', annot_author, results_folder)\n",
    "\n",
    "### Save labelling celltype1\n",
    "adata = bc.st.additional_labeling(adata, 'celltype2', 'celltype2', 'Intermediate cell types', annot_author, results_folder)\n",
    "\n",
    "\n",
    "### Save labelling dblabel used as reference\n",
    "adata = bc.st.additional_labeling(adata, 'dblabel', 'dblabel', 'Cell types attributed according to CL nomenclature, based on own annotation (celltype3)', annot_author, results_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Follow-up analyses for marker generation and inspecting the annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell-centered analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Breakdown of cell types per experiment (sample)\n",
    "bc.pl.celllabel_quant_stackedbar(adata, count_variable='celltype3', subset_variable = 'experiment');\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case two annotation versions are present, one can perfom a riverplot to compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bc.pl.riverplot_2categories(adata, ['dblabel','celltype3'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene-centered analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If one is interested in new markers, one can perform DE at the cell type annotation level of choice. Markers can then also be exported to a .gmtx file for subsequent import into GeMS (signatureDB), following specific conventions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if export_sigs:\n",
    "    ### Metadata setup for cell type signatures\n",
    "    User=author\n",
    "    Source='internal scseq'  \n",
    "    Subtype='all' # or healthy, onc, ...\n",
    "    domain='cell marker'\n",
    "    studyID='mongodb-study-id' # replace with your studyID\n",
    "    analysisID=analysis_name\n",
    "    genesetname=studyID+'_dblabel'\n",
    "    suffix='_model_user' # replace with specific info, e.g. _mc38_pcs\n",
    "    signature_dict={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform DE cells of each celltype3 vs. all other cells\n",
    "DEgenes=bc.tl.dge.get_de(adata,'celltype3',demethod='wilcoxon',topnr=5000, logfc=1,padj=0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Example 1: naive B cell markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Select only top 15 genes (in order of p-val) for 2 cell types and plot expression per cell type\n",
    "coi='naive B cell' # full dblabel\n",
    "coishort='NaiBcell' #dblabel short\n",
    "tops=list(DEgenes[coi].sort_values('Log2FC',ascending=False)['Name'][0:35])\n",
    "sc.pl.dotplot(adata, var_names=tops,groupby='celltype3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if export_sigs:\n",
    "    setName=coishort+suffix\n",
    "    desc='Genes higher expressed in '+ coi + ' vs. all other cells in MC38 in vivo exp ID tumor; coefs are log2FC'\n",
    "    pdout=DEgenes[coi].sort_values('Log2FC',ascending=False)[0:30]\n",
    "    genes=\"\\t\".join(list(pdout['Name'].astype(str) + \" | \" + pdout['Log2FC'].round(2).astype(str)))\n",
    "    signature_dict[setName] = bc.tl.sig.make_gmtx(setName,desc,User,Source,Subtype,domain,genesetname,\n",
    "                                                  genes,studyID,analysisID,celltype=coishort)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Example 2: pDC markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select only top 15 genes (in order of p-val) for 2 cell types and plot expression per cell type\n",
    "coi='plasmacytoid dendritic cell' # full dblabel\n",
    "coishort='pDC' #dblabel short\n",
    "tops=list(DEgenes[coi].sort_values('Log2FC',ascending=False)['Name'][0:35])\n",
    "sc.pl.dotplot(adata, var_names=tops,groupby='celltype3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if export_sigs:\n",
    "    setName=coishort+suffix\n",
    "    desc='Genes higher expressed in '+ coi + ' vs. all other cells in MC38 in vivo exp ID tumor; coefs are log2FC'\n",
    "    pdout=DEgenes[coi].sort_values('Log2FC',ascending=False)[0:30]\n",
    "    genes=\"\\t\".join(list(pdout['Name'].astype(str) + \" | \" + pdout['Log2FC'].round(2).astype(str)))\n",
    "    signature_dict[setName] = bc.tl.sig.make_gmtx(setName,desc,User,Source,Subtype,domain,genesetname,\n",
    "                                                  genes,studyID,analysisID,celltype=coishort)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Export sigs if relevant \n",
    "if export_sigs:\n",
    "    outgmtfile=results_folder+'/Celltypemarkers.gmtx'\n",
    "    bc.tl.sig.write_gmtx_forgems(signature_dict, outgmtfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If one is interested in additional markers correlated with the marker of interest, \n",
    "one can caculated the spearman correlation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as ss\n",
    "import operator\n",
    "\n",
    "allmarkers=adata.var_names\n",
    "\n",
    "spearcorsAll = {}\n",
    "for i in allmarkers:\n",
    "    spearcorsAll[i]=ss.spearmanr(adata.raw[:, i].X.todense(),adata.raw[:,'MS4A1'].X.todense())[0]\n",
    "\n",
    "\n",
    "spearcorsAll=sorted(spearcorsAll.items(),key=operator.itemgetter(1))\n",
    "spearcorsAll_ids = [idx for idx, val in spearcorsAll]\n",
    "goiNegAll=spearcorsAll_ids[0:30]\n",
    "goiPosAll=spearcorsAll_ids[(len(spearcorsAll_ids)-30):len(spearcorsAll_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, var_names=goiPosAll,groupby='celltype3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert to html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%javascript\n",
    "\n",
    "IPython.notebook.kernel.execute('nb_name = \"' + IPython.notebook.notebook_name + '\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nb_name = os.path.join(os.getcwd(), nb_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! jupyter nbconvert --to html {nb_name}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
