

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>recluster &mdash; BESCA 2.4.6+115.gb14a31c documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../_static/plot_directive.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-binder.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-dataframe.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sg_gallery-rendered-html.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="annotate_new_cellnames" href="besca.tl.rc.annotate_new_cellnames.html" />
    <link rel="prev" title="match_label" href="../sig/besca.tl.sig.match_label.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> BESCA
          

          
          </a>

          
            
            
              <div class="version">
                2.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../besca.html">besca</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../besca.html#helper-functions">helper functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../besca.html#preprocessing">preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../besca.html#plotting">plotting</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../besca.html#tools">tools</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.count_occurrence.html">count_occurrence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.count_occurrence_subset.html">count_occurrence_subset</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.count_occurrence_subset_conditions.html">count_occurrence_subset_conditions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.annotate_cells_clustering.html">annotate_cells_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.report.html">report</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools/besca.tl.plot_confusion_matrix.html">plot_confusion_matrix</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../besca.html#toolkits">toolkits</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../besca.html#batch-correction">batch correction</a></li>
<li class="toctree-l4"><a class="reference internal" href="../besca.html#differential-gene-expression">differential gene expression</a></li>
<li class="toctree-l4"><a class="reference internal" href="../besca.html#signature-scoring">signature scoring</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="../besca.html#reclustering">reclustering</a><ul class="current">
<li class="toctree-l5 current"><a class="current reference internal" href="#">recluster</a></li>
<li class="toctree-l5"><a class="reference internal" href="besca.tl.rc.annotate_new_cellnames.html">annotate_new_cellnames</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../besca.html#auto-annot">auto-annot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../besca.html#import">Import</a></li>
<li class="toctree-l2"><a class="reference internal" href="../besca.html#export">export</a></li>
<li class="toctree-l2"><a class="reference internal" href="../besca.html#standardworkflow">standardworkflow</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../auto_examples/index.html">code examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials.html">tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../besca_standard_pipeline.html">standard pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../adding_new_functions.html">adding new functions to besca</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">BESCA</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../besca.html">besca</a> &raquo;</li>
        
      <li>recluster</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/reclustering/besca.tl.rc.recluster.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="recluster">
<h1>recluster<a class="headerlink" href="#recluster" title="Permalink to this heading">Â¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="besca.tl.rc.recluster">
<span class="sig-prename descclassname"><span class="pre">besca.tl.rc.</span></span><span class="sig-name descname"><span class="pre">recluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltype_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0125</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_disp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regress_out_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_plot_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_shared</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/besca/tl/rc/_reclustering.html#recluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#besca.tl.rc.recluster" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Perform subclustering on specific celltype to identify subclusters.</p>
<p>Extract all cells that belong to the pre-labeled celltype into a new
data subset. This datasubset is initialized with the raw data contained in adata.raw. New highly
variable genes are selected and a new clustering is performed. The function returns the adata
subset with the new clustering annotation.</p>
<p>This can be performed on leiden clusters by setting celltype_label = âleidenâ and passing the
clusters that are to be selected for reclustering as strings or tuple of strings to the parameter
celltype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> â the complete AnnData object of the Dataset.</p></li>
<li><p><strong>celltype</strong> (<cite>str</cite> or (<cite>str</cite>)) â string identifying the cluster which is to be filtered out, if more than one is to be selected please
pass them as a tuple not as a list!</p></li>
<li><p><strong>celltype_label</strong> (<cite>str</cite> | default = âleidenâ) â string identifying which column in adata.obs will be matching with the celltype argument.</p></li>
<li><p><strong>min_mean</strong> (<cite>float</cite> | default = 0.0125) â the minimum gene expression a gene must have to be considered highly variable</p></li>
<li><p><strong>max_mean</strong> (<cite>float</cite> | default = 4) â the maximum gene expression a gene can have to be considered highly variable</p></li>
<li><p><strong>min_disp</strong> (<cite>float</cite> | default = 0.5) â the minimum dispersion a gene must have to be considered highly variable</p></li>
<li><p><strong>regress_out_key</strong> (<cite>list of str</cite> | default = None) â A list of string identifiers of the adata.obs columns that should be regressed out before
performing clustering. If None then no regress_out is calculated.</p></li>
<li><p><strong>random_seed</strong> (<cite>int</cite> | default = 0) â the random seed that is used to produce reproducible PCA, clustering and UMAP results</p></li>
<li><p><strong>show_plot_filter</strong> (<cite>bool</cite> | default = False) â boolian value indicating if a plot showing the filtering results for highly variable gene
detection should be displayed or not</p></li>
<li><p><strong>method</strong> (<cite>str</cite> | default = âleidenâ) â clustering method to use for the reclustering of the datasubset. Possible:louvain/leiden</p></li>
<li><p><strong>batch_key</strong> (<cite>str</cite> | default = None) â Specify a batch key if the HVG calculation should be done per batch</p></li>
<li><p><strong>n_share</strong> (<cite>int</cite> | default = 3) â Divide the nr. of batched by this nr. to get the shared HVGs considered (e.g. &gt;=1/3 of samples)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AnnData object containing the subcluster annotated with PCA, nearest neighbors, louvain cluster,</em></p></li>
<li><p><em>and UMAP coordinates.</em></p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>For a more detailed example of the entire reclustering process please refer to the code examples.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">besca</span> <span class="k">as</span> <span class="nn">bc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">simulated_pbmc3k_processed</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rc</span><span class="o">.</span><span class="n">recluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">celltype</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;6&#39;</span><span class="p">),</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">1.3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_subset</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_4&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_5&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_6&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_10&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_12&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene_20&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="besca.tl.rc.annotate_new_cellnames.html" class="btn btn-neutral float-right" title="annotate_new_cellnames" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../sig/besca.tl.sig.match_label.html" class="btn btn-neutral float-left" title="match_label" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, Roche Pharma Research and Early Development,     Pharmaceutical Sciences, Roche Innovation Center Basel, Basel, Switzerland.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>